// @cloudbase/weda-cloud-sdk/dist/h5/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccessToken = exports.initTcb = exports.clearCache = exports.dealCache = exports.getTcbInstance = exports.CLOUD_SDK = void 0;
const index_cover_1 = require("../index-cover");
const common_1 = require("../common");
const tcb_1 = require("./tcb");
const user_1 = __importDefault(require("./user"));
exports.CLOUD_SDK = Object.assign(index_cover_1.CLOUD_API, user_1.default);
// currentUser  为 getter  属性 , object.assign 会丢 getter 特性
// 在此 重新 设置上
Object.defineProperty(exports.CLOUD_SDK, 'currentUser', {
    get() {
        return user_1.default.currentUser;
    },
});
var tcb_2 = require("./tcb");
Object.defineProperty(exports, "getTcbInstance", { enumerable: true, get: function () { return tcb_2.getTcbInstance; } });
var cache_1 = require("./cache");
Object.defineProperty(exports, "dealCache", { enumerable: true, get: function () { return cache_1.dealCache; } });
Object.defineProperty(exports, "clearCache", { enumerable: true, get: function () { return cache_1.clearCache; } });
__exportStar(require("../index-cover"), exports);
/**
 * 初始化tcb并进行登录操作
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
async function _init(envID) {
    var _a, _b, _c, _d;
    const { app, auth } = await (0, tcb_1.getTcbInstance)();
    const customWedaConfig = (0, tcb_1.getCustomWedaConfig)();
    // 匿名登录/未登录均认为 未登录
    const loginState = await (0, tcb_1.getLoginState)();
    let isAnonymous = !loginState.hasLogin;
    /**
     * 1.在企业工作台中, 未具名登录
     * 2.编辑器实时预览超管登录,失败后走匿名
     */
    if ((customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.login) && (!loginState.hasLogin || (customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.forceLogin))) {
        try {
            await customWedaConfig.login({
                app,
                auth,
                loginState,
                async defaultLogin() {
                    await user_1.default.anonymousLogin();
                },
            });
            isAnonymous = false;
        }
        catch (e) {
            if (((_a = customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.loginConfig) === null || _a === void 0 ? void 0 : _a.failurePolicy) === 'anonymous_login') {
                await user_1.default.anonymousLogin();
                isAnonymous = true;
            }
            else {
                // 兼容旧异常逻辑处理
                throw e;
            }
        }
        // 未登录(非匿名也非具名登录)
    }
    else if (loginState.notLogin) {
        await user_1.default.anonymousLogin();
        isAnonymous = true;
    }
    // 处理后续事务: 注册用户(已注册的会跳过), 获取用户信息
    if (customWedaConfig && !((_b = customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.loginConfig) === null || _b === void 0 ? void 0 : _b.needSignIn)) {
        // 有用户角色信息则直接返回
        if (!((_c = customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.wedaUser) === null || _c === void 0 ? void 0 : _c.relatedRoles)) {
            await user_1.default.getUserInfo(customWedaConfig === null || customWedaConfig === void 0 ? void 0 : customWedaConfig.forceLogin);
        }
    }
    else {
        // 兼容jssdk1.0版本保持internalUser登录老逻辑，无internal标识都认为是外部用户
        const userType = isAnonymous
            ? 'anonymousUser'
            : loginState.version === '1.0' || ((_d = loginState.groups) === null || _d === void 0 ? void 0 : _d.includes('internal'))
                ? 'internalUser'
                : 'externalUser';
        try {
            await user_1.default.signIn({ userType });
        }
        catch (err) {
            console.warn('[weda-cloud-sdk] initial signIn failed', err);
        }
    }
    return {
        app,
        auth,
    };
}
async function initTcb(isForce = false) {
    const { envID } = (0, common_1.getConfig)();
    if (isForce) {
        return _init(envID);
    }
    return (0, common_1.execOnce)(_init, envID);
}
exports.initTcb = initTcb;
(0, index_cover_1.setConfig)({ initTcb });
/**
 * 获取 tcb accessToken
 */
async function getAccessToken() {
    const { auth } = await (0, tcb_1.getTcbInstance)();
    // @ts-ignore
    if (auth.getAccessToken) {
        // @ts-ignore
        return auth.getAccessToken();
    }
    if (auth.getAuthHeader) {
        const result = auth.getAuthHeader();
        if (!result)
            return result;
        return Object.assign({}, result, { accessToken: result['x-cloudbase-credentials'] });
    }
}
exports.getAccessToken = getAccessToken;
