// @cloudbase/weda-client/dist/cjs/web/actions/scanCode/scanCode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scanCode = void 0;
const react_1 = __importDefault(require("react"));
const react_dom_1 = __importDefault(require("react-dom"));
const ScanCodeComponent_1 = __importDefault(require("./ScanCodeComponent"));
const WEAPP_SCAN_CODE_ELEMENT_ID = 'weapp-scan-code-modal-root';
function scanCode(opts) {
    const options = {
        onlyFromCamera: false,
        scanType: ['barCode', 'qrCode'],
        success: () => { },
        fail: () => { },
        complete: () => { },
        enableDefaultBehavior: true,
        ...opts,
    };
    if (typeof options.scanType === 'string') {
        options.scanType = [options.scanType];
    }
    let root = document.getElementById(WEAPP_SCAN_CODE_ELEMENT_ID);
    if (!root) {
        root = document.createElement('div');
        root.id = WEAPP_SCAN_CODE_ELEMENT_ID;
    }
    document.body.appendChild(root);
    react_dom_1.default.render(react_1.default.createElement(ScanCodeComponent_1.default, { root: root, options: options }), root);
}
exports.scanCode = scanCode;
