// @tcwd/weapps-sdk/lib/flow.js
import { __assign } from "tslib";
import * as utils from './utils';
import * as kbone from 'kbone-api';
var events = {
    on: window.$$subscribe,
    off: window.$$unsubscribe,
    emit: window.$$publish,
};
function getFlowEventName(pageName) {
    return "weapps_flow_" + pageName;
}
function getPageName(url) {
    return url.split('/').slice(-1)[0];
}
function getFlowEventMap(pageName) {
    return window.$$global[getFlowEventName(pageName)];
}
// 创建流程
export function createFlow(url, options) {
    if (options === void 0) { options = {}; }
    var startPageLength = getCurrentPages().length;
    var pageName = getPageName(url);
    var successEvent = pageName + "-" + utils.generateGUID();
    var failEvent = pageName + "-" + utils.generateGUID();
    window.$$global[getFlowEventName(pageName)] = {
        successEvent: successEvent,
        failEvent: failEvent,
    };
    var urlWithOptions = utils.urlJoinParams(url, __assign({ successEvent: successEvent,
        failEvent: failEvent }, options));
    return new Promise(function (resolve, reject) {
        events.off(successEvent);
        events.off(failEvent);
        console.log('kbone flow 绑定事件', successEvent);
        events.on(successEvent, function (data) {
            console.log('-------------');
            console.log('流程成功接收到事件', successEvent, data, options);
            if (options.isKeepFlowPage) {
                // 保存到当前页面
                resolve(data);
            }
            else {
                // 如果是弹回流程开始页
                console.log('流程：', url);
                console.log('栈：', getCurrentPages());
                console.log('起始栈数：', startPageLength);
                console.log('当前栈数：', getCurrentPages().length);
                console.log('数据：', data);
                utils.navigateBack(getCurrentPages().length - startPageLength).then(function () {
                    resolve(data);
                });
            }
            console.log('-------------');
        });
        events.on(failEvent, function (err) { return reject(err); });
        console.log('跳转流程：', urlWithOptions);
        kbone
            .navigateTo({
            url: urlWithOptions,
        })
            .catch(function (e) { return console.error(e); });
    });
}
export function flowSuccess(data) {
    var currentRoute = getCurrentPages().slice(-1)[0].route;
    var pageName = getPageName(currentRoute);
    var successEvent = getFlowEventMap(pageName).successEvent;
    events.emit(successEvent, data);
}
export function flowFail(err) {
    var currentRoute = getCurrentPages().slice(-1)[0].route;
    var pageName = getPageName(currentRoute);
    var failEvent = getFlowEventMap(pageName).failEvent;
    events.emit(failEvent, err);
}
