// @tcwd/weapps-sdk/lib/i18n.js
var $weappsI18n = {
    langs: [],
    currentLang: '',
    i18nData: {},
    baseUrl: '',
};
var appNS = '$app';
var isWx = process.env.isMiniprogram;
export default {
    // 内部调用，开发者不要直接调
    __init: function (_a) {
        var langs = _a.langs, baseUrl = _a.baseUrl, _b = _a.i18nData, i18nData = _b === void 0 ? {} : _b, currentLang = _a.currentLang;
        return new Promise(function (resolve) {
            $weappsI18n.langs = langs;
            if (baseUrl)
                $weappsI18n.baseUrl = baseUrl;
            if (currentLang)
                $weappsI18n.currentLang = currentLang;
            if (i18nData) {
                $weappsI18n.i18nData = i18nData;
                resolve('');
            }
            else {
                // TODO: daniel 根据 baseUrl 去远程加载
            }
        });
    },
    __getI18nData: function () {
        return $weappsI18n.i18nData;
    },
    getLanguages: function () {
        return $weappsI18n.langs;
    },
    changeLanguage: function (code, autoReload) {
        if (autoReload === void 0) { autoReload = true; }
        // TODO: daniel暂时不实现从远程加载
        // $weappsI18n.i18nData 有数据，则加载 $weappsI18n.i18nData 的数据。没有则根据 baseUrl 从远程加载
        return new Promise(function (resolve) {
            $weappsI18n.currentLang = code;
            if (autoReload) {
                if (isWx) {
                    var pages = getCurrentPages();
                    var currentPage = pages[pages.length - 1];
                    var options_1 = currentPage.options || {};
                    var query = Object.keys(options_1)
                        .map(function (key) { return key + "=" + options_1[key]; })
                        .join('&');
                    wx.reLaunch({
                        url: "/" + currentPage.route + (query ? '?' + query : ''),
                    });
                }
                else {
                    localStorage.setItem('__select_language', code);
                    window.location.reload();
                }
            }
            else {
                resolve(code);
            }
        });
    },
    currentLanguage: function () {
        return $weappsI18n.currentLang;
    },
    t: function (key, data) {
        var ns = '';
        var currentLang = $weappsI18n.currentLang;
        var i18nData = $weappsI18n.i18nData[currentLang] || {};
        // 从key中提取出ns
        var matches = key.match(/(^[^:]*?):/);
        if (matches && matches.length >= 2) {
            ns = matches[1];
        }
        var langData;
        if (ns) {
            // key中有ns信息
            langData = i18nData[ns];
            if (langData) {
                // 这个 ns 信息是真的
                key = key.replace(new RegExp("^" + ns + ":"), ''); // 删掉ns才是真正的key
            }
        }
        if (!langData)
            langData = i18nData[appNS]; // 没有则默认取应用i18n信息
        return data
            ? ((langData && langData[key]) || key).replace(/({{(.*?)}})/g, function (replacement) {
                return data[replacement.replace(/{{(.*?)}}/, '$1')] || replacement;
            })
            : (langData && langData[key]) || key;
    },
};
