// @tcwd/weapps-sdk/lib/request.js
import { __assign } from "tslib";
import * as wxp from 'kbone-api';
import { getSessionId } from './session';
import { getConfig } from './config';
var showGlobalModal = false;
export default function request(options, retryTime) {
    if (retryTime === void 0) { retryTime = 2; }
    // 请求是否来自 weapps 系统
    var isWeApps = !!window.isWeApps;
    var config = getConfig();
    if (retryTime === 0) {
        return Promise.reject(null);
    }
    return getSessionId()
        .then(function (sessionid) {
        var _a;
        var _b;
        var config = getConfig();
        var configHeader = ((_b = config.service) === null || _b === void 0 ? void 0 : _b.headers) || {};
        var proxyHeader = {};
        var requestUrl = fixedDomain(options.url);
        if (isWeApps) {
            proxyHeader['weapps-proxy-url'] = requestUrl;
            requestUrl = location.origin + "/api/proxy";
        }
        options.header = options.header || {};
        var header = __assign(__assign(__assign(__assign({}, proxyHeader), options.header), (_a = { 'content-type': options.header['content-type'] || 'application/json' }, _a[config.sessionIdHeader] = sessionid, _a.appid = config.service.appId, _a)), configHeader);
        console.log('---------> weapps-sdk request begin', requestUrl, options, header);
        return wxp.request(__assign(__assign({}, options), { url: requestUrl, header: header, timeout: options.timeout || 15000 }));
    })
        .then(function (_a) {
        var statusCode = _a.statusCode, resp = _a.data;
        console.log('---------> weapps-sdk request finish', resp);
        wxp.hideLoading();
        var errcode = resp.errcode, code = resp.code;
        if (config.needReLoginErrorCode.includes(errcode)) {
            wxp.removeStorageSync(config.storageKeySessionId);
            return request(options, retryTime - 1);
        }
        if (statusCode === 200 && (Number(errcode) === 0 || Number(code) === 0)) {
            return resp.data;
        }
        else {
            console.error(resp.errmsg);
            throw resp;
        }
    })
        .catch(function (err) {
        if (!showGlobalModal && !config.ignoreErrorCode.includes(err.errcode)) {
            showGlobalModal = true;
            wxp.showModal({
                title: '温馨提示',
                content: '当前网络异常，请稍后再试',
                showCancel: false,
                success: function () {
                    showGlobalModal = false;
                },
            });
        }
        throw err;
    });
}
function fixedDomain(url) {
    var config = getConfig();
    if (url.indexOf('http') === 0) {
        return url;
    }
    return "" + config.service.domain + url;
}
