// @tcwd/weapps-sdk/lib/session.js
import { __assign } from "tslib";
import * as wxp from 'kbone-api';
import { getConfig } from './config';
var fetchSessionPromise;
var showGlobalModal = false;
function fetchSessionId() {
    // 请求是否来自 weapps 系统
    var isWeApps = !!process.env.isWeApps;
    var config = getConfig();
    if (fetchSessionPromise) {
        return fetchSessionPromise;
    }
    var _a = config.service, domain = _a.domain, getLoginUrl = _a.getLoginUrl, appId = _a.appId;
    fetchSessionPromise = wxp
        .login()
        .then(function (_a) {
        var code = _a.code;
        var proxyHeader = {};
        var requestUrl = domain + getLoginUrl(code);
        if (isWeApps) {
            // 转去 proxy 请求
            proxyHeader['weapps-proxy-url'] = requestUrl;
            requestUrl = location.origin + "/api/proxy";
        }
        return wxp.request({
            url: requestUrl,
            header: __assign(__assign({}, proxyHeader), { appid: appId }),
        });
    })
        .then(function (_a) {
        var data = _a.data;
        fetchSessionPromise = null;
        var errcode = data.errcode;
        if (errcode === 0) {
            var sessionid = data.data.sessionid || data.data.sessionId;
            var openid = data.data.openid || data.data.openId;
            localStorage.setItem(config.storageKeyOpenId, openid);
            localStorage.setItem(config.storageKeySessionId, sessionid);
            return sessionid;
        }
        throw handleSessionFail(data);
    })
        .catch(function (err) {
        fetchSessionPromise = null;
        throw handleSessionFail(err);
    });
    return fetchSessionPromise;
}
export function getSessionId() {
    var config = getConfig();
    var sessionid = localStorage.getItem(config.storageKeySessionId);
    if (sessionid) {
        return { then: function (cb) { return cb(sessionid); } };
    }
    return fetchSessionId();
}
function handleSessionFail(err) {
    var config = getConfig();
    if (!showGlobalModal && !config.ignoreErrorCode.includes(err.errcode)) {
        console.log('Get session error', err);
        showGlobalModal = true;
        wxp.showModal({
            title: '温馨提示',
            content: '当前网络异常，请稍后再试',
            showCancel: false,
            success: function () {
                showGlobalModal = false;
            },
        });
    }
    return {
        errcode: -20000,
        errmsg: '请求 sessionId 失败',
    };
}
