// @tcwd/weapps-sdk/lib/utils.js
import * as kbone from 'kbone-api';
// 处理url链接，加入params参数
export function urlJoinParams(url, params) {
    if (!url || !params || typeof params !== 'object') {
        return url;
    }
    var separate = url.indexOf('?') === -1 ? '?' : '&';
    var tempStr = Object.keys(params)
        .map(function (key) {
        if (typeof params[key] === 'object') {
            params[key] = JSON.stringify(params[key]);
        }
        if (params[key] !== undefined) {
            return key + "=" + encodeURIComponent(params[key]);
        }
        return '';
    })
        .filter(function (value) { return value; })
        .join('&');
    return "" + url + separate + tempStr;
}
// navigateBack
export function navigateBack(delta) {
    return new Promise(function (resolve) {
        var startDelta = getCurrentPages().length - delta;
        kbone.navigateBack({
            delta: delta,
            success: function () {
                var checkSuccess = function () {
                    setTimeout(function () {
                        var currentPageLength = getCurrentPages().length;
                        if (startDelta === currentPageLength || currentPageLength <= 1) {
                            console.log('已回退到指定页面：', getCurrentPages());
                            resolve();
                        }
                        else {
                            checkSuccess();
                        }
                    }, 100);
                };
                checkSuccess();
            },
        });
    });
}
// 生成唯一的 GUID
export function generateGUID() {
    var text = '';
    var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (var i = 0; i < 10; i += 1) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}
