// @cloudbase/weda-client/dist/cjs/mp/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._lodashSet = exports._lodashGet = exports.createComputed = exports.style2CssProperties = exports.px2rpx = exports.concatClassList = exports.watchAndSyncDatasetState2Local = exports.generateDatasetQuery = exports.createSimplePageAPIFromLaunchOptions = exports.PAGE_ROOT_SYMBOL = exports.createPage = exports.createComponent = exports.commonCompBehavior = exports._WEDA_CLOUD_SDK = exports.APP_LAUNCH_OPTIONS_PROMISE = exports.ControlledPromise = exports.urlJoinParams = exports.generatePageUrl = exports._getGeneralOptions = exports._ROUTER_KEY = exports._ACTIONS_KEY = exports.auth = exports.init = exports.getWedaAPI = exports.$w = exports.app = exports.createMpApp = void 0;
/**
 * 这里是对外暴露接口。必须严谨！！请勿随意对外暴露！！
 */
const DS_CLOUD_SDK = __importStar(require("@cloudbase/weda-cloud-sdk"));
const index_1 = require("./auth/index");
var app_1 = require("./app");
Object.defineProperty(exports, "createMpApp", { enumerable: true, get: function () { return app_1.createMpApp; } });
Object.defineProperty(exports, "app", { enumerable: true, get: function () { return app_1.app; } });
Object.defineProperty(exports, "$w", { enumerable: true, get: function () { return app_1.$w; } });
Object.defineProperty(exports, "getWedaAPI", { enumerable: true, get: function () { return app_1.getWedaAPI; } });
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return app_1.init; } });
var index_2 = require("./auth/index");
Object.defineProperty(exports, "auth", { enumerable: true, get: function () { return index_2.auth; } });
var constants_1 = require("../common/constants");
Object.defineProperty(exports, "_ACTIONS_KEY", { enumerable: true, get: function () { return constants_1.ACTIONS_KEY; } });
Object.defineProperty(exports, "_ROUTER_KEY", { enumerable: true, get: function () { return constants_1.ROUTER_KEY; } });
var index_3 = require("./models/index");
Object.defineProperty(exports, "_getGeneralOptions", { enumerable: true, get: function () { return index_3.getGeneralOptions; } });
var router_1 = require("./router");
Object.defineProperty(exports, "generatePageUrl", { enumerable: true, get: function () { return router_1.generatePageUrl; } });
var index_4 = require("../common/index");
Object.defineProperty(exports, "urlJoinParams", { enumerable: true, get: function () { return index_4.urlJoinParams; } });
Object.defineProperty(exports, "ControlledPromise", { enumerable: true, get: function () { return index_4.ControlledPromise; } });
Object.defineProperty(exports, "APP_LAUNCH_OPTIONS_PROMISE", { enumerable: true, get: function () { return index_4.APP_LAUNCH_OPTIONS_PROMISE; } });
/**
 * @private _前缀的都是内部方法，有可能调整，不推荐外部用户使用
 */
exports._WEDA_CLOUD_SDK = {
    setConfig: DS_CLOUD_SDK.setConfig,
    // initTcb: DS_CLOUD_SDK.initTcb,
    initTcb: async () => {
        await DS_CLOUD_SDK.initTcb();
        await index_1.auth.getUserInfo();
    },
    CLOUD_SDK: DS_CLOUD_SDK.CLOUD_SDK,
    createDataset: DS_CLOUD_SDK.createDataset,
    setLocalDatasetState: DS_CLOUD_SDK.setLocalDatasetState,
    createStateDataSourceVar: DS_CLOUD_SDK.createStateDataSourceVar,
    generateParamsParser: DS_CLOUD_SDK.generateParamsParser,
    setDatasetProfiles: DS_CLOUD_SDK.setDatasetProfiles,
    getDatasetProfiles: DS_CLOUD_SDK.getDatasetProfiles,
    EXTRA_API: DS_CLOUD_SDK.EXTRA_API,
    DS_SDK: DS_CLOUD_SDK.DS_SDK,
};
var behaviors_1 = require("./behaviors");
Object.defineProperty(exports, "commonCompBehavior", { enumerable: true, get: function () { return behaviors_1.commonCompBehavior; } });
var weapp_component_1 = require("./runtime/weapp-component");
Object.defineProperty(exports, "createComponent", { enumerable: true, get: function () { return weapp_component_1.createComponent; } });
var weapp_page_1 = require("./runtime/weapp-page");
Object.defineProperty(exports, "createPage", { enumerable: true, get: function () { return weapp_page_1.createPage; } });
Object.defineProperty(exports, "PAGE_ROOT_SYMBOL", { enumerable: true, get: function () { return weapp_page_1.PAGE_ROOT_SYMBOL; } });
Object.defineProperty(exports, "createSimplePageAPIFromLaunchOptions", { enumerable: true, get: function () { return weapp_page_1.createSimplePageAPIFromLaunchOptions; } });
var query_1 = require("./runtime/query");
Object.defineProperty(exports, "generateDatasetQuery", { enumerable: true, get: function () { return query_1.generateDatasetQuery; } });
var watch_1 = require("./runtime/watch");
Object.defineProperty(exports, "watchAndSyncDatasetState2Local", { enumerable: true, get: function () { return watch_1.watchAndSyncDatasetState2Local; } });
var style_1 = require("./runtime/style");
Object.defineProperty(exports, "concatClassList", { enumerable: true, get: function () { return style_1.concatClassList; } });
Object.defineProperty(exports, "px2rpx", { enumerable: true, get: function () { return style_1.px2rpx; } });
Object.defineProperty(exports, "style2CssProperties", { enumerable: true, get: function () { return style_1.style2CssProperties; } });
var util_1 = require("./runtime/util");
Object.defineProperty(exports, "createComputed", { enumerable: true, get: function () { return util_1.createComputed; } });
var vender_1 = require("../vender");
Object.defineProperty(exports, "_lodashGet", { enumerable: true, get: function () { return vender_1.get; } });
Object.defineProperty(exports, "_lodashSet", { enumerable: true, get: function () { return vender_1.set; } });
