// @cloudbase/weda-client/dist/cjs/web/app/tcb.js
"use strict";
/**
 * 初始化tcb，并挂载到_WedaHostConfig
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = void 0;
const weda_cloud_sdk_1 = require("@cloudbase/weda-cloud-sdk");
/**
 * 初始化工作台
 * @param params
 */
async function init(params) {
    (0, weda_cloud_sdk_1.setConfig)({
        isProd: params.isProd,
        envID: params.envId,
    });
    // const { app, auth } = await getTcbInstance(); // 不登录
    const { app, auth } = await (0, weda_cloud_sdk_1.getTcbInstance)(); // 匿名登录
    if (params?.type && params.type === 'portal') {
        // 企业工作台才需要挂载_WedaHostConfig
        await initWedaHostConfig({
            login: params.login,
            app,
            auth,
            type: params.type,
            envId: params.envId,
            isProd: params.isProd,
        });
    }
}
exports.init = init;
/**
 * 将tcb实例挂在到window上
 * @param params
 */
async function initWedaHostConfig(params) {
    window._WedaHostConfig = {
        // @ts-ignore 报错忽略，weda-cloud-sdk里面声明了
        appType: params.type,
        login: params.login,
        logout: params.logout,
        tcbInstance: params.app,
        envId: params.envId,
        isProd: params.isProd,
    };
}
