// @cloudbase/weda-client/dist/cjs/web/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.actionSdk = exports._WEDA_CLOUD_SDK = exports.APP_LAUNCH_OPTIONS_PROMISE = exports.ControlledPromise = exports.urlJoinParams = exports.generatePageUrl = exports._getGeneralOptions = exports._ROUTER_KEY = exports._ACTIONS_KEY = exports.auth = exports.getWedaAPI = exports.AppContext = exports.$w = exports.app = exports.createWebApp = void 0;
const DS_CLOUD_SDK = __importStar(require("@cloudbase/weda-cloud-sdk"));
/**
 * 这里是对外暴露接口。必须严谨！！请勿随意对外暴露！！
 */
var app_1 = require("./app");
Object.defineProperty(exports, "createWebApp", { enumerable: true, get: function () { return app_1.createWebApp; } });
Object.defineProperty(exports, "app", { enumerable: true, get: function () { return app_1.app; } });
Object.defineProperty(exports, "$w", { enumerable: true, get: function () { return app_1.$w; } });
Object.defineProperty(exports, "AppContext", { enumerable: true, get: function () { return app_1.AppContext; } });
Object.defineProperty(exports, "getWedaAPI", { enumerable: true, get: function () { return app_1.getWedaAPI; } });
var index_1 = require("./auth/index");
Object.defineProperty(exports, "auth", { enumerable: true, get: function () { return index_1.auth; } });
var constants_1 = require("../common/constants");
Object.defineProperty(exports, "_ACTIONS_KEY", { enumerable: true, get: function () { return constants_1.ACTIONS_KEY; } });
Object.defineProperty(exports, "_ROUTER_KEY", { enumerable: true, get: function () { return constants_1.ROUTER_KEY; } });
var models_1 = require("./models");
Object.defineProperty(exports, "_getGeneralOptions", { enumerable: true, get: function () { return models_1.getGeneralOptions; } });
var router_1 = require("./router");
Object.defineProperty(exports, "generatePageUrl", { enumerable: true, get: function () { return router_1.generatePageUrl; } });
var common_1 = require("../common");
Object.defineProperty(exports, "urlJoinParams", { enumerable: true, get: function () { return common_1.urlJoinParams; } });
Object.defineProperty(exports, "ControlledPromise", { enumerable: true, get: function () { return common_1.ControlledPromise; } });
Object.defineProperty(exports, "APP_LAUNCH_OPTIONS_PROMISE", { enumerable: true, get: function () { return common_1.APP_LAUNCH_OPTIONS_PROMISE; } });
/**
 * @private _前缀的都是内部方法，有可能调整，不推荐外部用户使用
 */
exports._WEDA_CLOUD_SDK = {
    setConfig: DS_CLOUD_SDK.setConfig,
    initTcb: DS_CLOUD_SDK.initTcb,
    CLOUD_SDK: DS_CLOUD_SDK.CLOUD_SDK,
    createDataset: DS_CLOUD_SDK.createDataset,
    setLocalDatasetState: DS_CLOUD_SDK.setLocalDatasetState,
    createStateDataSourceVar: DS_CLOUD_SDK.createStateDataSourceVar,
    generateParamsParser: DS_CLOUD_SDK.generateParamsParser,
    setDatasetProfiles: DS_CLOUD_SDK.setDatasetProfiles,
    getDatasetProfiles: DS_CLOUD_SDK.getDatasetProfiles,
    EXTRA_API: DS_CLOUD_SDK.EXTRA_API,
    DS_SDK: DS_CLOUD_SDK.DS_SDK,
};
var action_func_1 = require("./actions/action-func");
Object.defineProperty(exports, "actionSdk", { enumerable: true, get: function () { return action_func_1.actionSdk; } });
