// @cloudbase/node-sdk/lib/auth/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const utils_1 = require("../utils/utils");
const code_1 = require("../const/code");
const cloudbase_1 = require("../cloudbase");
const symbol_1 = require("../const/symbol");
const httpRequest_1 = __importDefault(require("../utils/httpRequest"));
const checkCustomUserIdRegex = /^[a-zA-Z0-9_\-#@~=*(){}[\]:.,<>+]{4,32}$/;
function validateUid(uid) {
    if (typeof uid !== 'string') {
        // console.log('debug:', { ...ERROR.INVALID_PARAM, message: 'uid must be a string' })
        throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: 'uid must be a string' }));
    }
    if (!checkCustomUserIdRegex.test(uid)) {
        throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: `Invalid uid: "${uid}"` }));
    }
}
function auth(cloudbase) {
    return {
        getUserInfo() {
            const { WX_OPENID, WX_APPID, TCB_UUID, TCB_CUSTOM_USER_ID, TCB_ISANONYMOUS_USER } = cloudbase_1.CloudBase.getCloudbaseContext();
            return {
                openId: WX_OPENID || '',
                appId: WX_APPID || '',
                uid: TCB_UUID || '',
                customUserId: TCB_CUSTOM_USER_ID || '',
                isAnonymous: TCB_ISANONYMOUS_USER === 'true' ? true : false
            };
        },
        getEndUserInfo(uid, opts) {
            const { WX_OPENID, WX_APPID, TCB_UUID, TCB_CUSTOM_USER_ID, TCB_ISANONYMOUS_USER } = cloudbase_1.CloudBase.getCloudbaseContext();
            const defaultUserInfo = {
                openId: WX_OPENID || '',
                appId: WX_APPID || '',
                uid: TCB_UUID || '',
                customUserId: TCB_CUSTOM_USER_ID || '',
                isAnonymous: TCB_ISANONYMOUS_USER === 'true' ? true : false
            };
            if (uid === undefined) {
                return {
                    userInfo: defaultUserInfo
                };
            }
            validateUid(uid);
            const params = {
                action: 'auth.getUserInfoForAdmin',
                uuid: uid
            };
            return httpRequest_1.default({
                config: cloudbase.config,
                params,
                method: 'post',
                opts,
                headers: {
                    'content-type': 'application/json'
                }
            }).then(res => {
                if (res.code) {
                    return res;
                }
                return {
                    userInfo: Object.assign({}, defaultUserInfo, res.data),
                    requestId: res.requestId
                };
            });
        },
        queryUserInfo(query, opts) {
            const { uid, platform, platformId } = query;
            const params = {
                action: 'auth.getUserInfoForAdmin',
                uuid: uid,
                platform,
                platformId
            };
            return httpRequest_1.default({
                config: cloudbase.config,
                params,
                method: 'post',
                opts,
                headers: {
                    'content-type': 'application/json'
                }
            }).then(res => {
                if (res.code) {
                    return res;
                }
                return {
                    userInfo: Object.assign({}, res.data),
                    requestId: res.requestId
                };
            });
        },
        async getAuthContext(context) {
            const { TCB_UUID, LOGINTYPE, QQ_OPENID, QQ_APPID } = cloudbase_1.CloudBase.getCloudbaseContext(context);
            const res = {
                uid: TCB_UUID,
                loginType: LOGINTYPE
            };
            if (LOGINTYPE === 'QQ-MINI') {
                res.appId = QQ_APPID;
                res.openId = QQ_OPENID;
            }
            return res;
        },
        getClientIP() {
            const { TCB_SOURCE_IP } = cloudbase_1.CloudBase.getCloudbaseContext();
            return TCB_SOURCE_IP || '';
        },
        createTicket: (uid, options = {}) => {
            validateUid(uid);
            const timestamp = new Date().getTime();
            const { TCB_ENV, SCF_NAMESPACE } = cloudbase_1.CloudBase.getCloudbaseContext();
            const { credentials } = cloudbase.config;
            const { env_id } = credentials;
            let { envName } = cloudbase.config;
            if (!envName) {
                throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: 'no env in config' }));
            }
            // æ£€æŸ¥credentials æ˜¯å¦åŒ…å«env
            if (!env_id) {
                throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: 'å½“å‰ç§é’¥æœªåŒ…å«env_id ä¿¡æ¯ï¼?è¯·å‰å¾€è…¾è®¯äº‘äº‘å¼€å‘æ§åˆ¶å°ï¼Œè·å–è‡ªå®šä¹‰ç™»å½•æœ€æ–°ç§é’? }));
            }
            // ä½¿ç”¨symbolæ—¶æ›¿æ¢ä¸ºç¯å¢ƒå˜é‡å†…çš„env
            if (envName === symbol_1.SYMBOL_CURRENT_ENV) {
                envName = TCB_ENV || SCF_NAMESPACE;
            }
            // æ£€æŸ?credentials env å’?init æŒ‡å®šenv æ˜¯å¦ä¸€è‡?            if (env_id && env_id !== envName) {
                throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: 'å½“å‰ç§é’¥æ‰€å±ç¯å¢ƒä¸ init æŒ‡å®šç¯å¢ƒä¸ä¸€è‡´ï¼' }));
            }
            const { refresh = 3600 * 1000, expire = timestamp + 7 * 24 * 60 * 60 * 1000 } = options;
            const token = jsonwebtoken_1.default.sign({
                alg: 'RS256',
                env: envName,
                iat: timestamp,
                exp: timestamp + 10 * 60 * 1000,
                uid,
                refresh,
                expire
            }, credentials.private_key, { algorithm: 'RS256' });
            return credentials.private_key_id + '/@@/' + token;
        }
    };
}
exports.auth = auth;
