// @cloudbase/oauth/dist/esm/captcha/captcha-dom.js
import { CloudbaseEventEmitter } from '@cloudbase/utilities/dist/cjs/libs/events';
import { parseCaptcha } from '@cloudbase/utilities/dist/cjs/libs/util';
const debounce = (func, delay) => {
    let timeoutId;
    return (...args) => {
        clearTimeout(timeoutId);
        timeoutId = setTimeout(() => func.apply(null, args), delay);
    };
};
const eventBus = new CloudbaseEventEmitter();
const CAPTCHA_EVENT = {
    CAPTCHA_DATA_CHANGE: 'captchaDataChange',
    RESOLVE_CAPTCHA_DATA: 'resolveCaptchaData',
};
const CAPTCHA_STYLES = {
    overlay: 'position:fixed;top:0;left:0;width:100%;height:100%;background:rgba(0,0,0,0.5);z-index:999',
    modal: 'position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);width:320px;padding:20px;background:#fff;border-radius:8px;box-shadow:0 0 10px rgba(0,0,0,0.2);z-index:1000',
    prompt: 'margin-bottom:15px',
    captchaWrap: 'display:flex;align-items:center;justify-content:space-between;margin-bottom:15px',
    captchaImg: 'width:100%;height:auto;border:1px solid #eee',
    refreshBtn: 'padding:10px 8px;background:#007bff;color:#fff;border:none;border-radius:4px;cursor:pointer;flex:1;white-space:nowrap;margin-left:10px',
    input: 'width:100%;padding:14px 16px;margin-bottom:15px;box-sizing:border-box;border:2px solid #e8ecf4;border-radius:10px;font-size:15px',
    confirmBtn: 'width:100%;padding:10px;background:#007bff;color:#fff;border:none;border-radius:4px;cursor:pointer',
    loading: 'background:#6c757d;cursor:not-allowed',
};
const genCaptchaDom = ({ oauthInstance, captchaState }) => {
    const CAPTCHA_IMG_ID = 'captcha-image';
    const refreshCaptcha = async () => {
        try {
            const result = await oauthInstance.createCaptchaData({ state: captchaState.state });
            captchaState = { ...captchaState, captchaData: result.data, token: result.token };
            document.getElementById(CAPTCHA_IMG_ID).src = result.data;
        }
        catch (error) {
            console.error('刷新图形验证码失败', error);
        }
    };
    const createElement = (tag, styles, text) => {
        const el = document.createElement(tag);
        if (styles)
            el.style.cssText = styles;
        if (text)
            el.textContent = text;
        return el;
    };
    const overlay = createElement('div', CAPTCHA_STYLES.overlay);
    const modal = createElement('div', CAPTCHA_STYLES.modal);
    modal.appendChild(createElement('p', CAPTCHA_STYLES.prompt, '请输入你看到的字符：'));
    const captchaWrap = createElement('div', CAPTCHA_STYLES.captchaWrap);
    const captchaImg = createElement('img', CAPTCHA_STYLES.captchaImg);
    captchaImg.id = CAPTCHA_IMG_ID;
    captchaImg.src = captchaState.captchaData;
    captchaWrap.appendChild(captchaImg);
    const refreshBtn = createElement('button', CAPTCHA_STYLES.refreshBtn, '刷新');
    refreshBtn.onclick = debounce(async () => {
        refreshBtn.textContent = '加载中...';
        refreshBtn.disabled = true;
        refreshBtn.style.cssText = `${CAPTCHA_STYLES.refreshBtn};${CAPTCHA_STYLES.loading}`;
        try {
            await refreshCaptcha();
        }
        finally {
            refreshBtn.textContent = '刷新';
            refreshBtn.disabled = false;
            refreshBtn.style.cssText = CAPTCHA_STYLES.refreshBtn;
        }
    }, 500);
    captchaWrap.appendChild(refreshBtn);
    modal.appendChild(captchaWrap);
    const input = createElement('input', CAPTCHA_STYLES.input);
    input.type = 'text';
    input.placeholder = '输入字符';
    input.maxLength = 4;
    modal.appendChild(input);
    const errorMsg = createElement('div', 'color:#dc3545;font-size:12px;margin-bottom:10px;display:none;padding:8px;background:#f8d7da;border:1px solid #f5c6cb;border-radius:4px');
    modal.appendChild(errorMsg);
    const showError = (message) => {
        errorMsg.textContent = message;
        errorMsg.style.display = 'block';
    };
    const confirmBtn = createElement('button', CAPTCHA_STYLES.confirmBtn, '确定');
    confirmBtn.onclick = debounce(async () => {
        const inputValue = input.value.trim();
        if (!inputValue) {
            showError('请输入字符');
            return;
        }
        errorMsg.style.display = 'none';
        confirmBtn.textContent = '验证中...';
        confirmBtn.disabled = true;
        confirmBtn.style.cssText = `${CAPTCHA_STYLES.confirmBtn};${CAPTCHA_STYLES.loading}`;
        try {
            const verifyResult = await oauthInstance.verifyCaptchaData({
                token: captchaState.token,
                key: inputValue,
            });
            eventBus.fire(CAPTCHA_EVENT.RESOLVE_CAPTCHA_DATA, verifyResult);
            console.log('图形验证码校验成功');
            document.body.removeChild(overlay);
            document.body.removeChild(modal);
        }
        catch (error) {
            console.error('图形验证码校验失败', error);
            const errorMessage = error.error_description || '图形验证码校验失败';
            showError(errorMessage);
            input.value = '';
            input.focus();
            await refreshCaptcha();
        }
        finally {
            confirmBtn.textContent = '确定';
            confirmBtn.disabled = false;
            confirmBtn.style.cssText = CAPTCHA_STYLES.confirmBtn;
        }
    }, 500);
    modal.appendChild(confirmBtn);
    document.body.appendChild(overlay);
    document.body.appendChild(modal);
};
export const openURIWithCallback = async (url, oauthInstance) => {
    const { captchaData, state, token } = parseCaptcha(url);
    genCaptchaDom({ oauthInstance, captchaState: { captchaData, state, token } });
    return new Promise((resolve) => {
        console.log('等待图形验证码校验结果...');
        eventBus.on(CAPTCHA_EVENT.RESOLVE_CAPTCHA_DATA, (res) => {
            resolve(res?.data);
        });
    });
};
