// @tcwd/weapps-sdk/lib/app-h5-sdk/router/index.js
import { urlJoinParams } from '../../';
import { getConfig, preHashPath } from '../setConfig';
function routerGo(path, action) {
    if (action === void 0) { action = 'push'; }
    if (process.env.isApp) {
        location.hash = path;
    }
    else {
        window._WEAPPS_HISTORY[action](path);
    }
}

function generatePathname(path, indexPage) {
  if(!/^\//.test(path) && window._WEAPPS_HISTORY && window._WEAPPS_HISTORY.location && window._WEAPPS_HISTORY.location.pathname && /\/$/.test(window._WEAPPS_HISTORY.location.pathname)) {
    const suffix = path && !/\/$/.test(path) ? '/': ''
    return '/' + path + suffix
  }
  if(/^\//.test(path) || /^\./.test(path)){
    return path
  }
  return "/" + path
}

export function redirectTo(_a) {
    var pageId = _a.pageId, packageName = _a.packageName, params = _a.params, indexPage = _a.indexPage;
    var url = generatePathname(packageName ? packageName + "/" + pageId : "" + pageId, indexPage);
    if (!process.env.isApp) {
        var path = urlJoinParams(url, params);
        routerGo(path, 'replace');
    }
}
export function navigateTo(_a) {
    var pageId = _a.pageId, packageName = _a.packageName, params = _a.params, events = _a.events, success = _a.success, fail = _a.fail, complete = _a.complete, indexPage = _a.indexPage;
    var url = generatePathname(packageName ? packageName + "/" + pageId : "" + pageId, indexPage);
    if (!process.env.isApp) {
        var path = urlJoinParams(url, params);
        routerGo(path, 'push');
    }
    else {
        var _b = getConfig().tabBar, tabBar = _b === void 0 ? {} : _b;
        var list = tabBar.list;
        var isInTabBar_1 = false;
        list.forEach(function (item) {
            var pagePath = item.pagePath.replace(preHashPath, '');
            if (url === pagePath) {
                console.warn(url + " \u5DF2\u8BBE\u7F6E\u5728tabBar\u91CC \u4E0D\u53EF\u8DF3\u8F6C");
                isInTabBar_1 = true;
            }
        });
        if (isInTabBar_1) {
            return;
        }
        console.warn(url + " \u4E0D\u5728tabBar\u91CC\uFF0C\u53EF\u4EE5\u8DF3\u8F6C...");
        var path = urlJoinParams(url, params);
        routerGo(path);
    }
}
export function reLaunch(_a) {
    var pageId = _a.pageId, packageName = _a.packageName, params = _a.params, events = _a.events, success = _a.success, fail = _a.fail, complete = _a.complete, indexPage = _a.indexPage;
    var url = generatePathname(packageName ? packageName + "/" + pageId : "" + pageId, indexPage);
    var path = urlJoinParams(url, params);
    routerGo(path, 'replace');
}
export function switchTab(_a) {
    var url = _a.url;
    var paths = url.split('?');
    var hash = paths[0];
    var result = /pages\/([0-9a-zA-z_]+)\/index/.exec(url);
    if (result.length > 1) {
        hash = "" + result[1];
    }
    if (!process.env.isApp) {
        routerGo("" + hash + (paths[1] ? "?" + paths[1] : ''));
    }
    else {
        var _b = getConfig().tabBar, tabBar = _b === void 0 ? {} : _b;
        var list = tabBar.list;
        var isInTabBar_2 = false;
        list.forEach(function (item) {
            var pagePath = item.pagePath.replace(preHashPath, '');
            if (hash === pagePath) {
                isInTabBar_2 = true;
            }
        });
        if (isInTabBar_2) {
            console.log("switchTab " + url + " \u5DF2\u8BBE\u7F6E\u5728tabBar\u91CC");
            routerGo("" + hash + (paths[1] ? "?" + paths[1] : ''));
        }
    }
}
