// @tcwd/weapps-sdk/lib/app-h5-sdk/h5Api/index.js
/**
 * app 中 使用web 能力的API， 在web端和app里加入
 */
import { fromByteArray, toByteArray } from 'base64-js';
import queryString from "querystring";
import { createAsyncFunc } from '../utils/tools';
export { pageScrollTo } from '../scroll';
export { default as chooseLocation } from '../location/chooseLocation';
export { createSelectorQuery } from '../createSelectorQuery';
// export { uploadFile, downloadFile } from '../fileTransfer-native'
// export { getRecorderManager } from '../recorderManager-native'
// 做tree-shaking按需引入
export var getRecorderManager;
export var uploadFile;
export var downloadFile;
if (process.env.isApp) {
    getRecorderManager = require('../recorderManager-native').getRecorderManager;
    uploadFile = require('../fileTransfer-native').uploadFile;
    downloadFile = require('../fileTransfer-native').downloadFile;
}
export { connectSocket, onSocketOpen, onSocketError, sendSocketMessage, onSocketMessage, closeSocket, onSocketClose } from '../webSocket';
export { showToast, hideToast, showLoading, hideLoading, showModal, showActionSheet } from '../interactive';
export { setConfig } from '../setConfig';
export { navigateTo, switchTab, reLaunch, redirectTo } from '../router';
export function arrayBufferToBase64(arrayBuffer) {
    return fromByteArray(arrayBuffer);
}
export function base64ToArrayBuffer(base64) {
    return toByteArray(base64);
}
export function navigateBack(json) {
    if (json === void 0) { json = {}; }
    var delta = -(~~json.delta || 1);
    history.go(delta);
}
export function nextTick(cb) {
    setTimeout(cb, 0);
}
export function onError(cb) {
    window.addEventListener('error', cb, false);
}
export function getSetting(json) {
    if (json === void 0) { json = {}; }
    return createAsyncFunc(json, { authSetting: {} }, 'err');
}
export function getLaunchOptionsSync() {
    var query = queryString.parse(location.search.replace('?', ''));
    return {
        path: location.hash,
        query: query,
        referrerInfo: {},
        scene: undefined,
        shareTicket: undefined
    };
}
export var env = {};
export { default as os } from '../os';
export { default as request } from '../request';
