// @cloudbase/weda-client/dist/esm/web/app/tcb.js
/**
 * 初始化tcb，并挂载到_WedaHostConfig
 */
import { setConfig, getTcbInstance } from '@cloudbase/weda-cloud-sdk';
/**
 * 初始化工作台
 * @param params
 */
export async function init(params) {
    setConfig({
        isProd: params.isProd,
        envID: params.envId,
    });
    // const { app, auth } = await getTcbInstance(); // 不登录
    const { app, auth } = await getTcbInstance(); // 匿名登录
    if (params?.type && params.type === 'portal') {
        // 企业工作台才需要挂载_WedaHostConfig
        await initWedaHostConfig({
            login: params.login,
            app,
            auth,
            type: params.type,
            envId: params.envId,
            isProd: params.isProd,
        });
    }
}
/**
 * 将tcb实例挂在到window上
 * @param params
 */
async function initWedaHostConfig(params) {
    window._WedaHostConfig = {
        // @ts-ignore 报错忽略，weda-cloud-sdk里面声明了
        appType: params.type,
        login: params.login,
        logout: params.logout,
        tcbInstance: params.app,
        envId: params.envId,
        isProd: params.isProd,
    };
}
