// @cloudbase/weda-client/dist/cjs/common/logic/declare.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    If: {
        id: 'If',
        type: 'method',
        name: '条件分支',
        funName: 'If',
        editorGram: 'If(${1:true}, ${2:"您好"}, ${3:"hello"})',
        gram: 'If(判断条件, 满足时返回的值, 不满足时返回的值)',
        desc: '按判断条件进行逻辑比较，满足时返回一个值，不满足时返回另一个值。',
        cases: '判断 1 是否等于 1，正确返回"您好"，错误返回"hello"，结果返回"您好"。',
        example: 'If(1==1, "您好", "hello")',
        schemaRetType: { type: 'any' },
        retType: 'any',
        Zindex: 0,
    },
    IsEmpty: {
        id: 'IsEmpty',
        type: 'method',
        name: '是否为空',
        funName: 'IsEmpty',
        gram: 'IsEmpty(值)',
        editorGram: 'IsEmpty(${1:"hello"})',
        desc: '判断输入值是否为空，如果结果为数组类型，还会判断结果数组项是否全为空。',
        cases: '判断变量text是否为空文本，如果是空返回true，反之返回false',
        example: 'IsEmpty(text)',
        schemaRetType: { type: 'boolean' },
        retType: 'boolean',
        Zindex: 1,
    },
    NotNull: {
        id: 'NotNull',
        type: 'method',
        name: '是否为null',
        funName: 'NotNull',
        gram: 'NotNull(字段)',
        editorGram: 'NotNull(${1:null})',
        desc: '判断传入的字段值是否为null',
        cases: '判断变量val是否为null，如果是null返回true，反之返回false',
        example: 'NotNull(val)',
        schemaRetType: { type: 'boolean' },
        retType: 'boolean',
        Zindex: 2,
    },
    And: {
        id: 'And',
        type: 'method',
        name: '与',
        funName: 'And',
        gram: 'And(判断条件1, [判断条件2,...])',
        editorGram: 'And(${1:true}, ${2:false})',
        desc: '用于确定所有判断条件是否为真',
        cases: '判断 1 是否等于 1，2 是否等于 2，如果全部为真返回true，反之返回false',
        example: 'And(1==1, 2==2)',
        schemaRetType: { type: 'boolean' },
        retType: 'boolean',
        Zindex: 3,
    },
    Or: {
        id: 'Or',
        type: 'method',
        name: '或',
        funName: 'Or',
        gram: 'Or(判断条件1, [判断条件2,...])',
        editorGram: 'Or(${1:true}, ${2:false})',
        desc: '任意一个判断条件为真，则结果为真；所有条件为否，结果才为否',
        cases: '判断 1 是否等于 1，2 是否等于 2，如果任意为真返回true，反之返回false',
        example: 'Or(1==1, 2==2)',
        schemaRetType: { type: 'boolean' },
        retType: 'boolean',
        Zindex: 4,
    },
};
