// @tcwd/weapps-sdk/lib/app-h5-sdk/setConfig/index.js
import { __assign } from "tslib";
var __weappsNative = window.__weappsNative;
export var preHashPath = '/';
var appConfig = {};
export var setConfig = function (config) {
    if (process.env.isApp) {
        try {
            var _a = config.generate, generate = _a === void 0 ? {} : _a, _b = config.appExtraConfig, appExtraConfig = _b === void 0 ? {} : _b, __homePath__ = config.__homePath__, app = config.app, pages_1 = config.pages;
            var pagePaths = Object.keys(pages_1);
            // pages添加pagePath，以及样式设置
            var __pages = pagePaths.map(function (path) {
                return __assign(__assign({}, pages_1[path].extra), { pagePath: "" + preHashPath + [path] });
            });
            var tabBar = generate.tabBar || appExtraConfig.tabBar || {};
            var _c = tabBar.list, list = _c === void 0 ? [] : _c;
            tabBar.list = list.map(function (item) {
                var pagePath = item.pagePath;
                var result = /pages\/([0-9a-zA-z_]+)\/index/.exec(pagePath);
                if (result.length > 1) {
                    pagePath = "" + preHashPath + result[1];
                }
                return __assign(__assign({}, item), { pagePath: pagePath });
            });
            var data = {
                home: preHashPath + __homePath__,
                window: app,
                pages: __pages,
                tabBar: tabBar
            };
            var params = {
                data: data
            };
            appConfig = data;
            console.log("__weappsNative.setConfig \u5F00\u59CB\u6267\u884C...", JSON.stringify(params));
            __weappsNative && __weappsNative.setConfig(JSON.stringify(params));
        }
        catch (e) {
            console.log(e);
        }
    }
};
export var getConfig = function () {
    return appConfig;
};
