// @cloudbase/weda-client/dist/cjs/mp/auth/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.auth = void 0;
const user_1 = require("./user");
const uitl_1 = require("./uitl");
const weda_cloud_sdk_1 = require("@cloudbase/weda-cloud-sdk");
/**
 * @description 获取权限、登录相关信息的对象
 */
exports.auth = {
    signIn: user_1.signIn,
    signOut: user_1.signOut,
    getUserInfo: user_1.getUserInfo,
    getAccessToken: user_1.getAccessToken,
    loginScope: user_1.loginScope,
    checkAnonymous: uitl_1.checkAnonymous,
    get currentUser() {
        return user_1.userObj.currentUser;
    },
    openIdLoginInWxApp: weda_cloud_sdk_1.CLOUD_SDK.openIdLoginInWxApp,
    unionIdLoginInWxApp: weda_cloud_sdk_1.CLOUD_SDK.unionIdLoginInWxApp,
    modifyCurrentUser: weda_cloud_sdk_1.CLOUD_SDK.modifyCurrentUser,
    __internal__: {
        redirectToLogin: uitl_1.redirectToLogin,
        getAuthConfig: uitl_1.getAuthConfig,
        getAccessPermission: uitl_1.getAccessPermission,
        findLoginPage: uitl_1.findLoginPage,
        cleanAccessPermissionCache: uitl_1.cleanAccessPermissionCache,
    },
};
