// @cloudbase/auth/dist/cjs/adapter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useAuthAdapter = void 0;
var utilities_1 = require("./utilities");
var OauthClientStorageBase = (function () {
    function OauthClientStorageBase(config) {
        var _a, _b;
        try {
            this.localStorage = (config === null || config === void 0 ? void 0 : config.localStorage) || (((_b = (_a = config === null || config === void 0 ? void 0 : config.adapter) === null || _a === void 0 ? void 0 : _a.root) === null || _b === void 0 ? void 0 : _b.globalThis) || globalThis).localStorage;
        }
        catch (error) {
            this.localStorage = (config === null || config === void 0 ? void 0 : config.localStorage) || {};
        }
    }
    OauthClientStorageBase.prototype.getItem = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.localStorage.getItem(key)];
            });
        });
    };
    OauthClientStorageBase.prototype.removeItem = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.localStorage.removeItem(key);
                return [2];
            });
        });
    };
    OauthClientStorageBase.prototype.setItem = function (key, value) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.localStorage.setItem(key, value);
                return [2];
            });
        });
    };
    OauthClientStorageBase.prototype.getItemSync = function (key) {
        return this.localStorage.getItem(key);
    };
    OauthClientStorageBase.prototype.setItemSync = function (key, value) {
        this.localStorage.setItem(key, value);
    };
    OauthClientStorageBase.prototype.removeItemSync = function (key) {
        this.localStorage.removeItem(key);
    };
    return OauthClientStorageBase;
}());
var URL_REG = /^[^:]+:\/\/[^/]+(\/[^?#]+)/;
var dealAdapterRequestData = function (res, options) {
    var _a, _b, _c;
    var requestId = (_a = options === null || options === void 0 ? void 0 : options.headers) === null || _a === void 0 ? void 0 : _a['x-request-id'];
    if (!res) {
        res = {};
    }
    if ((res === null || res === void 0 ? void 0 : res.code) || ((_b = res === null || res === void 0 ? void 0 : res.data) === null || _b === void 0 ? void 0 : _b.error_code)) {
        res = {
            error: res.code || res.data.error,
            error_description: res.data.error_description || res.message || res.code || res.data.error_code,
            request_id: res.requestId,
            error_code: (_c = res.data) === null || _c === void 0 ? void 0 : _c.error_code,
        };
    }
    if (!(res === null || res === void 0 ? void 0 : res.request_id)) {
        res.request_id = res.request_id || requestId;
    }
    if (res === null || res === void 0 ? void 0 : res.error) {
        res.error_uri = getUrlPath(options === null || options === void 0 ? void 0 : options.url);
        return res;
    }
    return (res === null || res === void 0 ? void 0 : res.data) || {};
};
function generateOauthClientRequest(reqClass) {
    return function (url, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, new Promise(function (resolve, reject) {
                        var copyOptions = Object.assign({}, options);
                        if (copyOptions.body && typeof copyOptions.body !== 'string') {
                            copyOptions.body = JSON.stringify(copyOptions.body);
                        }
                        var task = wx.request({
                            url: url,
                            data: copyOptions.body,
                            timeout: reqClass._timeout,
                            method: copyOptions.method || 'GET',
                            header: copyOptions.headers,
                            success: function (res) {
                                res = dealAdapterRequestData(res, __assign(__assign({}, options), { url: url }));
                                if (res.error) {
                                    reject(res);
                                }
                                resolve(res);
                            },
                            fail: function (err) {
                                reject({
                                    error: 'unreachable',
                                    error_description: err.message,
                                });
                            },
                            complete: function (err) {
                                if (!err || !err.errMsg) {
                                    return;
                                }
                                if (!reqClass._timeout || reqClass._restrictedMethods.indexOf('request') === -1) {
                                    return;
                                }
                                var errMsg = err.errMsg;
                                if (errMsg === 'request:fail timeout') {
                                    console.warn(reqClass._timeoutMsg);
                                    try {
                                        task.abort();
                                    }
                                    catch (e) { }
                                }
                            },
                        });
                    })];
            });
        });
    };
}
function getUrlPath(url) {
    return URL_REG.test(url) ? RegExp.$1 : url;
}
var dealUserAdapter = function (config, adapter) {
    var data = __assign({}, config);
    try {
        if (!(config === null || config === void 0 ? void 0 : config.storage) && (adapter === null || adapter === void 0 ? void 0 : adapter.localStorage)) {
            data.storage = new OauthClientStorageBase({ localStorage: adapter.localStorage, adapter: adapter });
        }
        if (!config.captchaOptions && (adapter === null || adapter === void 0 ? void 0 : adapter.captchaOptions)) {
            data.captchaOptions = adapter.captchaOptions;
        }
        if (!config.baseRequest && !config.request && (adapter === null || adapter === void 0 ? void 0 : adapter.reqClass)) {
            var reqClassParams = {
                timeout: 10000,
                restrictedMethods: ['get', 'post', 'upload', 'download', 'request'],
            };
            var adapterReq_1 = new adapter.reqClass(reqClassParams);
            data.baseRequest = function (url, options) {
                var _a, _b, _c, _d, _e;
                return __awaiter(this, void 0, void 0, function () {
                    var res, requestType, func, error_1;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                res = {};
                                requestType = ((_a = options.method) === null || _a === void 0 ? void 0 : _a.toLocaleLowerCase()) || 'post';
                                func = (_c = (_b = adapterReq_1 === null || adapterReq_1 === void 0 ? void 0 : adapterReq_1[requestType]) === null || _b === void 0 ? void 0 : _b.bind) === null || _c === void 0 ? void 0 : _c.call(_b, adapterReq_1);
                                if (typeof func !== 'function') {
                                    throw new Error("invalid \u3010".concat(requestType, "\u3011 request method"));
                                }
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                if ((!((_d = options.headers) === null || _d === void 0 ? void 0 : _d['content-type']) || ((_e = options.headers) === null || _e === void 0 ? void 0 : _e['content-type'].includes('application/json'))) && !['string', 'undefined'].includes(typeof options.body)) {
                                    options.data = options.data || options.body;
                                    options.body = JSON.stringify(options.body);
                                }
                                return [4, func(__assign({ url: url }, options))];
                            case 2:
                                res = _f.sent();
                                res = dealAdapterRequestData(res, __assign(__assign({}, options), { url: url }));
                                return [3, 4];
                            case 3:
                                error_1 = _f.sent();
                                res = __assign(__assign({}, error_1 === null || error_1 === void 0 ? void 0 : error_1.data), error_1);
                                throw res;
                            case 4:
                                if (res === null || res === void 0 ? void 0 : res.error) {
                                    throw res;
                                }
                                return [2, res];
                        }
                    });
                });
            };
        }
    }
    catch (e) {
        console.error('user adapter generate fail:', e);
    }
    return data;
};
var dealAuthAdapter = function (config, adapter) {
    var data = __assign({}, config);
    var authOptions = {};
    if (adapter === null || adapter === void 0 ? void 0 : adapter.captchaOptions) {
        data.captchaOptions = adapter.captchaOptions;
    }
    if (!utilities_1.adapterForWxMp.isMatch() || (config.storage && config.baseRequest))
        return data;
    try {
        var _a = utilities_1.adapterForWxMp.genAdapter({}), localStorage_1 = _a.localStorage, ReqClass = _a.reqClass;
        if (localStorage_1) {
            authOptions.storage = new OauthClientStorageBase({ localStorage: localStorage_1, adapter: adapter });
        }
        if (ReqClass) {
            var reqClass = new ReqClass({
                timeout: 10000,
                restrictedMethods: ['get', 'post', 'upload', 'download', 'request'],
            });
            authOptions.request = generateOauthClientRequest(reqClass);
        }
        if (data.captchaOptions) {
            authOptions.baseRequest = authOptions.request;
            authOptions.request = undefined;
        }
        return __assign(__assign({}, data), authOptions);
    }
    catch (e) {
        console.error('adapter generate fail:', e);
    }
    return data;
};
var useAuthAdapter = function (config) {
    if (config.adapter && config.adapter.type !== 'default') {
        return dealUserAdapter(config, config.adapter);
    }
    return dealAuthAdapter(config, config.adapter);
};
exports.useAuthAdapter = useAuthAdapter;
