// @cloudbase/auth/dist/cjs/sbaseApi.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SbaseApi = exports.isBrowser = void 0;
var oauth_1 = require("@cloudbase/oauth");
var utils_1 = require("./utils");
var utilities_1 = require("@cloudbase/utilities");
var utilities_2 = require("./utilities");
var isBrowser = function () { return typeof window !== 'undefined' && typeof document !== 'undefined'; };
exports.isBrowser = isBrowser;
var SbaseApi = (function () {
    function SbaseApi(config) {
        this.listeners = new Map();
        this.hasListenerSetUp = false;
        this.config = config;
        this.oauthInstance = config.oauthInstance;
        this.cache = config.cache;
        this.init();
    }
    Object.defineProperty(SbaseApi.prototype, "currentUser", {
        get: function () {
            var loginState = this.hasLoginState();
            if (loginState) {
                return loginState.user || null;
            }
            return null;
        },
        enumerable: false,
        configurable: true
    });
    SbaseApi.prototype.hasLoginState = function () {
        throw new Error('Auth.hasLoginState() is not implemented');
    };
    SbaseApi.prototype.setAccessKey = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.setAccessKey() is not implemented');
            });
        });
    };
    SbaseApi.prototype.signIn = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.signIn() is not implemented');
            });
        });
    };
    SbaseApi.prototype.signInWithProvider = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.signInWithProvider() is not implemented');
            });
        });
    };
    SbaseApi.prototype.genProviderRedirectUri = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.genProviderRedirectUri() is not implemented');
            });
        });
    };
    SbaseApi.prototype.getAccessToken = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.getAccessToken() is not implemented');
            });
        });
    };
    SbaseApi.prototype.getUserInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.getUserInfo() is not implemented');
            });
        });
    };
    SbaseApi.prototype.updateUserBasicInfo = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.updateUserBasicInfo() is not implemented');
            });
        });
    };
    SbaseApi.prototype.setCustomSignFunc = function (_getTickFn) {
        throw new Error('Auth.setCustomSignFunc() is not implemented');
    };
    SbaseApi.prototype.grantProviderToken = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.grantProviderToken() is not implemented');
            });
        });
    };
    SbaseApi.prototype.bindWithProvider = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.bindWithProvider() is not implemented');
            });
        });
    };
    SbaseApi.prototype.signInWithUsername = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.signInWithUsername() is not implemented');
            });
        });
    };
    SbaseApi.prototype.getVerification = function (_params, _options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.getVerification() is not implemented');
            });
        });
    };
    SbaseApi.prototype.createLoginState = function (_params, _options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.createLoginState() is not implemented');
            });
        });
    };
    SbaseApi.prototype.verify = function (_params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Auth.verify() is not implemented');
            });
        });
    };
    SbaseApi.prototype.signInAnonymously = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var loginState, _a, _b, session, error_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 4, , 5]);
                        return [4, this.oauthInstance.authApi.signInAnonymously(params)];
                    case 1:
                        _c.sent();
                        return [4, this.createLoginState()];
                    case 2:
                        loginState = _c.sent();
                        return [4, this.getSession()];
                    case 3:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 4:
                        error_1 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_1) }];
                    case 5: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signUp = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var verificationInfo_1, error_2;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(params.phone_number || params.verification_code || params.verification_token || params.provider_token)) return [3, 2];
                        return [4, this.oauthInstance.authApi.signUp(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        return [4, this.getVerification(params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })];
                    case 3:
                        verificationInfo_1 = _a.sent();
                        return [2, {
                                data: {
                                    verifyOtp: function (_a) {
                                        var token = _a.token, _b = _a.messageId, messageId = _b === void 0 ? verificationInfo_1.verification_id : _b;
                                        return __awaiter(_this, void 0, void 0, function () {
                                            var verificationTokenRes, data, _c, _d, session, error_3;
                                            return __generator(this, function (_e) {
                                                switch (_e.label) {
                                                    case 0:
                                                        _e.trys.push([0, 8, , 9]);
                                                        return [4, this.verify({
                                                                verification_id: messageId || verificationInfo_1.verification_id,
                                                                verification_code: token,
                                                            })];
                                                    case 1:
                                                        verificationTokenRes = _e.sent();
                                                        if (!verificationInfo_1.is_user) return [3, 3];
                                                        return [4, this.signIn({
                                                                username: params.email || this.formatPhone(params.phone),
                                                                verification_token: verificationTokenRes.verification_token,
                                                            })];
                                                    case 2:
                                                        _e.sent();
                                                        return [3, 6];
                                                    case 3:
                                                        data = JSON.parse(JSON.stringify(params));
                                                        delete data.email;
                                                        delete data.phone;
                                                        return [4, this.oauthInstance.authApi.signUp(__assign(__assign(__assign({}, data), (params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })), { verification_token: verificationTokenRes.verification_token, verification_code: token }))];
                                                    case 4:
                                                        _e.sent();
                                                        return [4, this.createLoginState()];
                                                    case 5:
                                                        _e.sent();
                                                        _e.label = 6;
                                                    case 6: return [4, this.getSession()];
                                                    case 7:
                                                        _c = (_e.sent()).data, _d = _c === void 0 ? {} : _c, session = _d.session;
                                                        return [2, { data: { user: session.user, session: session }, error: null }];
                                                    case 8:
                                                        error_3 = _e.sent();
                                                        return [2, { data: {}, error: new oauth_1.AuthError(error_3) }];
                                                    case 9: return [2];
                                                }
                                            });
                                        });
                                    },
                                },
                                error: null,
                            }];
                    case 4:
                        error_2 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_2) }];
                    case 5: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signOut = function (params) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var userInfoKey, res, error_4;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        userInfoKey = this.cache.keys.userInfoKey;
                        return [4, this.oauthInstance.authApi.signOut(params)];
                    case 1:
                        res = _c.sent();
                        return [4, this.cache.removeStoreAsync(userInfoKey)];
                    case 2:
                        _c.sent();
                        this.setAccessKey();
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.LOGIN_STATE_CHANGED, { eventType: oauth_1.LOGIN_STATE_CHANGED_TYPE.SIGN_OUT });
                        (_b = this.config.eventBus) === null || _b === void 0 ? void 0 : _b.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_OUT });
                        return [2, __assign(__assign({}, res), { data: {}, error: null })];
                    case 3:
                        error_4 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_4) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.onAuthStateChange = function (callback) {
        var _this = this;
        if (!this.hasListenerSetUp) {
            this.setupListeners();
            this.hasListenerSetUp = true;
        }
        var id = Math.random().toString(36);
        if (!this.listeners.has(id)) {
            this.listeners.set(id, new Set());
        }
        this.listeners.get(id).add(callback);
        var subscription = {
            id: id,
            callback: callback,
            unsubscribe: function () {
                var callbacks = _this.listeners.get(id);
                if (callbacks) {
                    callbacks.delete(callback);
                    if (callbacks.size === 0) {
                        _this.listeners.delete(id);
                    }
                }
            },
        };
        return {
            data: { subscription: subscription },
        };
    };
    SbaseApi.prototype.signInWithPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_5;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        this.validateAtLeastOne(params, [['username'], ['email'], ['phone']], 'You must provide either username, email, or phone');
                        this.validateParams(params, {
                            password: { required: true, message: 'Password is required' },
                        });
                        return [4, this.signIn(__assign({ username: params.username || params.email || this.formatPhone(params.phone), password: params.password }, (params.is_encrypt ? { isEncrypt: true, version: 'v2' } : {})))];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_5 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_5) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithIdToken = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_6;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            token: { required: true, message: 'Token is required' },
                        });
                        return [4, this.signInWithProvider({
                                provider_token: params.token,
                            })];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_6 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_6) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithOtp = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var verificationInfo_2, error_7;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        return [4, this.getVerification(params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })];
                    case 1:
                        verificationInfo_2 = _a.sent();
                        return [2, {
                                data: {
                                    user: null,
                                    session: null,
                                    verifyOtp: function (_a) {
                                        var token = _a.token, _b = _a.messageId, messageId = _b === void 0 ? verificationInfo_2.verification_id : _b;
                                        return __awaiter(_this, void 0, void 0, function () {
                                            return __generator(this, function (_c) {
                                                return [2, this.verifyOtp({
                                                        email: params.email,
                                                        phone: params.phone,
                                                        token: token,
                                                        messageId: messageId,
                                                    })];
                                            });
                                        });
                                    },
                                },
                                error: null,
                            }];
                    case 2:
                        error_7 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_7) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.verifyOAuth = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var data, code, state, cacheData, provider, token, res, localSearch, error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        data = {};
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 7, , 8]);
                        code = (params === null || params === void 0 ? void 0 : params.code) || utilities_1.utils.getQuery('code');
                        state = (params === null || params === void 0 ? void 0 : params.state) || utilities_1.utils.getQuery('state');
                        if (!code) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'Code is required' }) }];
                        }
                        if (!state) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'State is required' }) }];
                        }
                        cacheData = (0, utils_1.getBrowserSession)(state);
                        data.type = cacheData === null || cacheData === void 0 ? void 0 : cacheData.type;
                        provider = (params === null || params === void 0 ? void 0 : params.provider) || (cacheData === null || cacheData === void 0 ? void 0 : cacheData.provider) || utilities_1.utils.getQuery('provider');
                        if (!provider) {
                            return [2, { data: data, error: new oauth_1.AuthError({ message: 'Provider is required' }) }];
                        }
                        return [4, this.grantProviderToken({
                                provider_id: provider,
                                provider_redirect_uri: location.origin + location.pathname,
                                provider_code: code,
                            })];
                    case 2:
                        token = (_a.sent()).provider_token;
                        res = void 0;
                        if (!(cacheData.type === oauth_1.OAUTH_TYPE.BIND_IDENTITY)) return [3, 4];
                        return [4, this.oauthInstance.authApi.toBindIdentity({ provider_token: token, provider: provider, fireEvent: true })];
                    case 3:
                        res = _a.sent();
                        return [3, 6];
                    case 4: return [4, this.signInWithIdToken({
                            token: token,
                        })];
                    case 5:
                        res = _a.sent();
                        res.data = __assign(__assign({}, data), res.data);
                        _a.label = 6;
                    case 6:
                        localSearch = new URLSearchParams(location === null || location === void 0 ? void 0 : location.search);
                        localSearch.delete('code');
                        localSearch.delete('state');
                        (0, utils_1.addUrlSearch)((cacheData === null || cacheData === void 0 ? void 0 : cacheData.search) === undefined ? "?".concat(localSearch.toString()) : cacheData === null || cacheData === void 0 ? void 0 : cacheData.search, (cacheData === null || cacheData === void 0 ? void 0 : cacheData.hash) || location.hash);
                        (0, utils_1.removeBrowserSession)(state);
                        return [2, res];
                    case 7:
                        error_8 = _a.sent();
                        return [2, { data: data, error: new oauth_1.AuthError(error_8) }];
                    case 8: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithOAuth = function (params) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var href, urlObject, provider_redirect_uri, state, uri, decodedUri, finalUri, url_1, error_9;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        href = ((_a = params.options) === null || _a === void 0 ? void 0 : _a.redirectTo) || location.href;
                        urlObject = new URL(href);
                        provider_redirect_uri = urlObject.origin + urlObject.pathname;
                        state = ((_b = params.options) === null || _b === void 0 ? void 0 : _b.state) || "prd-".concat(params.provider, "-").concat(Math.random().toString(36)
                            .slice(2));
                        return [4, this.genProviderRedirectUri({
                                provider_id: params.provider,
                                provider_redirect_uri: provider_redirect_uri,
                                state: state,
                            })];
                    case 1:
                        uri = (_f.sent()).uri;
                        decodedUri = decodeURIComponent(uri);
                        finalUri = decodedUri;
                        if ((_c = params.options) === null || _c === void 0 ? void 0 : _c.queryParams) {
                            url_1 = new URL(decodedUri);
                            Object.entries(params.options.queryParams).forEach(function (_a) {
                                var key = _a[0], value = _a[1];
                                url_1.searchParams.set(key, value);
                            });
                            finalUri = url_1.toString();
                        }
                        (0, utils_1.saveToBrowserSession)(state, {
                            provider: params.provider,
                            search: urlObject.search,
                            hash: urlObject.hash,
                            type: ((_d = params.options) === null || _d === void 0 ? void 0 : _d.type) || oauth_1.OAUTH_TYPE.SIGN_IN,
                        });
                        if ((0, exports.isBrowser)() && !((_e = params.options) === null || _e === void 0 ? void 0 : _e.skipBrowserRedirect)) {
                            window.location.assign(finalUri);
                        }
                        return [2, { data: { url: finalUri, provider: params.provider }, error: null }];
                    case 2:
                        error_9 = _f.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_9) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithSSO = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2];
            });
        });
    };
    SbaseApi.prototype.signInWithWeb3 = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2];
            });
        });
    };
    SbaseApi.prototype.getClaims = function () {
        return __awaiter(this, void 0, void 0, function () {
            var accessToken, parsedToken, error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4, this.getAccessToken()];
                    case 1:
                        accessToken = (_a.sent()).accessToken;
                        parsedToken = (0, oauth_1.weappJwtDecodeAll)(accessToken);
                        return [2, { data: parsedToken, error: null }];
                    case 2:
                        error_10 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_10) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.resetPasswordForEmail = function (emailOrPhone, options) {
        return __awaiter(this, void 0, void 0, function () {
            var redirectTo_1, isEmail_1, verificationParams, formattedPhone, verificationInfo_3, error_11;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams({ emailOrPhone: emailOrPhone }, {
                            emailOrPhone: { required: true, message: 'Email or phone is required' },
                        });
                        redirectTo_1 = (options || {}).redirectTo;
                        isEmail_1 = emailOrPhone.includes('@');
                        verificationParams = void 0;
                        if (isEmail_1) {
                            verificationParams = { email: emailOrPhone };
                        }
                        else {
                            formattedPhone = this.formatPhone(emailOrPhone);
                            verificationParams = { phone_number: formattedPhone };
                        }
                        return [4, this.getVerification(verificationParams)];
                    case 1:
                        verificationInfo_3 = _a.sent();
                        return [2, {
                                data: {
                                    updateUser: function (attributes) { return __awaiter(_this, void 0, void 0, function () {
                                        var verificationTokenRes, res, error_12;
                                        var _a;
                                        return __generator(this, function (_b) {
                                            switch (_b.label) {
                                                case 0:
                                                    this.validateParams(attributes, {
                                                        nonce: { required: true, message: 'Nonce is required' },
                                                        password: { required: true, message: 'Password is required' },
                                                    });
                                                    _b.label = 1;
                                                case 1:
                                                    _b.trys.push([1, 5, , 6]);
                                                    return [4, this.verify({
                                                            verification_id: verificationInfo_3.verification_id,
                                                            verification_code: attributes.nonce,
                                                        })];
                                                case 2:
                                                    verificationTokenRes = _b.sent();
                                                    return [4, this.oauthInstance.authApi.resetPassword({
                                                            email: isEmail_1 ? emailOrPhone : undefined,
                                                            phone: !isEmail_1 ? emailOrPhone : undefined,
                                                            new_password: attributes.password,
                                                            verification_token: verificationTokenRes.verification_token,
                                                        })];
                                                case 3:
                                                    _b.sent();
                                                    (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.PASSWORD_RECOVERY });
                                                    return [4, this.signInWithPassword({
                                                            email: isEmail_1 ? emailOrPhone : undefined,
                                                            phone: !isEmail_1 ? emailOrPhone : undefined,
                                                            password: attributes.password,
                                                        })];
                                                case 4:
                                                    res = _b.sent();
                                                    if (redirectTo_1 && (0, exports.isBrowser)()) {
                                                        window.location.assign(redirectTo_1);
                                                    }
                                                    return [2, res];
                                                case 5:
                                                    error_12 = _b.sent();
                                                    return [2, { data: {}, error: new oauth_1.AuthError(error_12) }];
                                                case 6: return [2];
                                            }
                                        });
                                    }); },
                                },
                                error: null,
                            }];
                    case 2:
                        error_11 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_11) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.resetPasswordForOld = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_13;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.oauthInstance.authApi.updatePasswordByOld({
                                old_password: params.old_password,
                                new_password: params.new_password,
                            })];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_13 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_13) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.verifyOtp = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var type, _a, _b, session, error_14;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 6, , 7]);
                        type = params.type;
                        this.validateParams(params, {
                            token: { required: true, message: 'Token is required' },
                            messageId: { required: true, message: 'messageId is required' },
                        });
                        if (!['phone_change', 'email_change'].includes(type)) return [3, 2];
                        return [4, this.verify({
                                verification_id: params.messageId,
                                verification_code: params.token,
                            })];
                    case 1:
                        _c.sent();
                        return [3, 4];
                    case 2: return [4, this.signInWithUsername({
                            verificationInfo: { verification_id: params.messageId, is_user: true },
                            verificationCode: params.token,
                            username: params.email || this.formatPhone(params.phone) || '',
                            loginType: params.email ? 'email' : 'phone',
                        })];
                    case 3:
                        _c.sent();
                        _c.label = 4;
                    case 4: return [4, this.getSession()];
                    case 5:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 6:
                        error_14 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_14) }];
                    case 7: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.getSession = function () {
        return __awaiter(this, void 0, void 0, function () {
            var credentials, _a, _b, user, error_15;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.oauthInstance.oauth2client.getCredentials()];
                    case 1:
                        credentials = _c.sent();
                        if (!credentials || credentials.scope === 'accessKey') {
                            return [2, { data: { session: null }, error: null }];
                        }
                        return [4, this.getUser()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, user = _b.user;
                        return [2, { data: { session: __assign(__assign({}, credentials), { user: user }) }, error: null }];
                    case 3:
                        error_15 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_15) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.refreshSession = function (refresh_token) {
        return __awaiter(this, void 0, void 0, function () {
            var credentials, newTokens, _a, _b, user, error_16;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 4, , 5]);
                        return [4, this.oauthInstance.oauth2client.localCredentials.getCredentials()];
                    case 1:
                        credentials = _c.sent();
                        credentials.refresh_token = refresh_token || credentials.refresh_token;
                        return [4, this.oauthInstance.oauth2client.refreshToken(credentials)];
                    case 2:
                        newTokens = _c.sent();
                        return [4, this.getUser()];
                    case 3:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, user = _b.user;
                        return [2, { data: { user: user, session: __assign(__assign({}, newTokens), { user: user }) }, error: null }];
                    case 4:
                        error_16 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_16) }];
                    case 5: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.getUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            var user, _a, error_17;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        _a = this.convertToUser;
                        return [4, this.getUserInfo()];
                    case 1:
                        user = _a.apply(this, [_b.sent()]);
                        return [2, { data: { user: user }, error: null }];
                    case 2:
                        error_17 = _b.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_17) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.refreshUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_18;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.currentUser.refresh()];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_18 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_18) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.updateUser = function (params) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var hasValue, email, phone, restParams, needsEmailVerification_1, needsPhoneVerification, extraRes, verificationParams, verificationType_1, formattedPhone, verificationInfo_4, _b, user, error_19;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 8, , 9]);
                        hasValue = Object.keys(params).some(function (key) { return params[key] !== undefined && params[key] !== null && params[key] !== ''; });
                        if (!hasValue) {
                            throw new oauth_1.AuthError({ message: 'At least one field must be provided for update' });
                        }
                        email = params.email, phone = params.phone, restParams = __rest(params, ["email", "phone"]);
                        needsEmailVerification_1 = email !== undefined;
                        needsPhoneVerification = phone !== undefined;
                        extraRes = {};
                        if (!(needsEmailVerification_1 || needsPhoneVerification)) return [3, 4];
                        verificationParams = void 0;
                        if (needsEmailVerification_1) {
                            verificationParams = { email: params.email };
                            verificationType_1 = 'email_change';
                        }
                        else {
                            formattedPhone = this.formatPhone(params.phone);
                            verificationParams = { phone_number: formattedPhone };
                            verificationType_1 = 'phone_change';
                        }
                        return [4, this.getVerification(verificationParams)];
                    case 1:
                        verificationInfo_4 = _c.sent();
                        _b = Object.keys(restParams).length > 0;
                        if (!_b) return [3, 3];
                        return [4, this.updateUserBasicInfo(restParams)];
                    case 2:
                        _b = (_c.sent());
                        _c.label = 3;
                    case 3:
                        _b;
                        extraRes = {
                            messageId: verificationInfo_4.verification_id,
                            verifyOtp: function (verifyParams) { return __awaiter(_this, void 0, void 0, function () {
                                var user_1, error_20;
                                var _a;
                                return __generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            _b.trys.push([0, 11, , 12]);
                                            if (!(verifyParams.email && params.email === verifyParams.email)) return [3, 3];
                                            return [4, this.verifyOtp({
                                                    type: 'email_change',
                                                    email: params.email,
                                                    token: verifyParams.token,
                                                    messageId: verificationInfo_4.verification_id,
                                                })];
                                        case 1:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo({ email: params.email })];
                                        case 2:
                                            _b.sent();
                                            return [3, 9];
                                        case 3:
                                            if (!(verifyParams.phone && params.phone === verifyParams.phone)) return [3, 6];
                                            return [4, this.verifyOtp({
                                                    type: 'phone_change',
                                                    phone: params.phone,
                                                    token: verifyParams.token,
                                                    messageId: verificationInfo_4.verification_id,
                                                })];
                                        case 4:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo({ phone: this.formatPhone(params.phone) })];
                                        case 5:
                                            _b.sent();
                                            return [3, 9];
                                        case 6: return [4, this.verifyOtp({
                                                type: verificationType_1,
                                                email: needsEmailVerification_1 ? params.email : undefined,
                                                phone: !needsEmailVerification_1 ? params.phone : undefined,
                                                token: verifyParams.token,
                                                messageId: verificationInfo_4.verification_id,
                                            })];
                                        case 7:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo(params)];
                                        case 8:
                                            _b.sent();
                                            _b.label = 9;
                                        case 9: return [4, this.getUser()];
                                        case 10:
                                            user_1 = (_b.sent()).data.user;
                                            (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.USER_UPDATED });
                                            return [2, { data: { user: user_1 }, error: null }];
                                        case 11:
                                            error_20 = _b.sent();
                                            return [2, { data: {}, error: new oauth_1.AuthError(error_20) }];
                                        case 12: return [2];
                                    }
                                });
                            }); },
                        };
                        return [3, 6];
                    case 4: return [4, this.updateUserBasicInfo(params)];
                    case 5:
                        _c.sent();
                        _c.label = 6;
                    case 6: return [4, this.getUser()];
                    case 7:
                        user = (_c.sent()).data.user;
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.USER_UPDATED });
                        return [2, { data: __assign({ user: user }, extraRes), error: null }];
                    case 8:
                        error_19 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_19) }];
                    case 9: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.getUserIdentities = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var providers, error_21;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4, this.oauthInstance.authApi.getProviders()];
                    case 1:
                        providers = _b.sent();
                        return [2, { data: { identities: (_a = providers === null || providers === void 0 ? void 0 : providers.data) === null || _a === void 0 ? void 0 : _a.filter(function (v) { return !!v.bind; }) }, error: null }];
                    case 2:
                        error_21 = _b.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_21) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.linkIdentity = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var error_22;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        return [4, this.signInWithOAuth({
                                provider: params.provider,
                                options: {
                                    type: oauth_1.OAUTH_TYPE.BIND_IDENTITY,
                                },
                            })];
                    case 1:
                        _a.sent();
                        return [2, { data: { provider: params.provider }, error: null }];
                    case 2:
                        error_22 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_22) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.unlinkIdentity = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var error_23;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        return [4, this.oauthInstance.authApi.unbindProvider({ provider_id: params.provider })];
                    case 1:
                        _a.sent();
                        return [2, { data: {}, error: null }];
                    case 2:
                        error_23 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_23) }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.reauthenticate = function () {
        return __awaiter(this, void 0, void 0, function () {
            var user, userInfo_1, verificationInfo_5, error_24;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4, this.getUser()];
                    case 1:
                        user = (_a.sent()).data.user;
                        this.validateAtLeastOne(user, [['email', 'phone']], 'You must provide either an email or phone number');
                        userInfo_1 = user.email ? { email: user.email } : { phone_number: this.formatPhone(user.phone) };
                        return [4, this.getVerification(userInfo_1)];
                    case 2:
                        verificationInfo_5 = _a.sent();
                        return [2, {
                                data: {
                                    updateUser: function (attributes) { return __awaiter(_this, void 0, void 0, function () {
                                        var verificationTokenRes, sudoRes, _a, _b, session, error_25;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    this.validateParams(attributes, {
                                                        nonce: { required: true, message: 'Nonce is required' },
                                                    });
                                                    _c.label = 1;
                                                case 1:
                                                    _c.trys.push([1, 9, , 10]);
                                                    if (!attributes.password) return [3, 5];
                                                    return [4, this.verify({
                                                            verification_id: verificationInfo_5.verification_id,
                                                            verification_code: attributes.nonce,
                                                        })];
                                                case 2:
                                                    verificationTokenRes = _c.sent();
                                                    return [4, this.oauthInstance.authApi.sudo({
                                                            verification_token: verificationTokenRes.verification_token,
                                                        })];
                                                case 3:
                                                    sudoRes = _c.sent();
                                                    return [4, this.oauthInstance.authApi.setPassword({
                                                            new_password: attributes.password,
                                                            sudo_token: sudoRes.sudo_token,
                                                        })];
                                                case 4:
                                                    _c.sent();
                                                    return [3, 7];
                                                case 5: return [4, this.signInWithUsername(__assign(__assign({ verificationInfo: verificationInfo_5, verificationCode: attributes.nonce }, userInfo_1), { loginType: userInfo_1.email ? 'email' : 'phone' }))];
                                                case 6:
                                                    _c.sent();
                                                    _c.label = 7;
                                                case 7: return [4, this.getSession()];
                                                case 8:
                                                    _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                                                    return [2, { data: { user: session.user, session: session }, error: null }];
                                                case 9:
                                                    error_25 = _c.sent();
                                                    return [2, { data: {}, error: new oauth_1.AuthError(error_25) }];
                                                case 10: return [2];
                                            }
                                        });
                                    }); },
                                },
                                error: null,
                            }];
                    case 3:
                        error_24 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_24) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.resend = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var target, data, verificationId, error_26;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        target = params.type === 'signup' ? 'ANY' : 'USER';
                        data = { target: target };
                        if ('email' in params) {
                            data.email = params.email;
                        }
                        if ('phone' in params) {
                            data.phone_number = this.formatPhone(params.phone);
                        }
                        return [4, this.oauthInstance.authApi.getVerification(data)];
                    case 1:
                        verificationId = (_a.sent()).verification_id;
                        return [2, {
                                data: { messageId: verificationId },
                                error: null,
                            }];
                    case 2:
                        error_26 = _a.sent();
                        return [2, {
                                data: {},
                                error: new oauth_1.AuthError(error_26),
                            }];
                    case 3: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.setSession = function (params) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b, _c, session, error_27;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            access_token: { required: true, message: 'Access token is required' },
                            refresh_token: { required: true, message: 'Refresh token is required' },
                        });
                        return [4, this.oauthInstance.oauth2client.refreshToken(params, { throwOnError: true })];
                    case 1:
                        _d.sent();
                        return [4, this.getSession()];
                    case 2:
                        _b = (_d.sent()).data, _c = _b === void 0 ? {} : _b, session = _c.session;
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN });
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_27 = _d.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_27) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.exchangeCodeForSession = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2];
            });
        });
    };
    SbaseApi.prototype.deleteUser = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var sudo_token, error_28;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            password: { required: true, message: 'Password is required' },
                        });
                        return [4, this.oauthInstance.authApi.sudo(params)];
                    case 1:
                        sudo_token = (_a.sent()).sudo_token;
                        return [4, this.oauthInstance.authApi.deleteMe({ sudo_token: sudo_token })];
                    case 2:
                        _a.sent();
                        return [2, { data: {}, error: null }];
                    case 3:
                        error_28 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_28) }];
                    case 4: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.toDefaultLoginPage = function (params) {
        if (params === void 0) { params = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var configVersion, query, redirectUri, urlObj, loginPage;
            return __generator(this, function (_a) {
                try {
                    configVersion = params.config_version || 'env';
                    query = Object.keys(params.query || {})
                        .map(function (key) { return "".concat(key, "=").concat(params.query[key]); })
                        .join('&');
                    if (utilities_2.adapterForWxMp.isMatch()) {
                        wx.navigateTo({ url: "/packages/$wd_system/pages/login/index".concat(query ? "?".concat(query) : '') });
                    }
                    else {
                        redirectUri = params.redirect_uri || window.location.href;
                        urlObj = new URL(redirectUri);
                        loginPage = "".concat(urlObj.origin, "/__auth/?app_id=").concat(params.app_id || '', "&env_id=").concat(this.config.env, "&client_id=").concat(this.config.clientId || this.config.env, "&config_version=").concat(configVersion, "&redirect_uri=").concat(encodeURIComponent(redirectUri)).concat(query ? "&".concat(query) : '');
                        window.location.href = loginPage;
                    }
                    return [2, { data: {}, error: null }];
                }
                catch (error) {
                    return [2, { data: {}, error: new oauth_1.AuthError(error) }];
                }
                return [2];
            });
        });
    };
    SbaseApi.prototype.signInWithCustomTicket = function (getTickFn) {
        return __awaiter(this, void 0, void 0, function () {
            var loginState, _a, _b, session, error_29;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (getTickFn) {
                            this.setCustomSignFunc(getTickFn);
                        }
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 5, , 6]);
                        return [4, this.oauthInstance.authApi.signInWithCustomTicket()];
                    case 2:
                        _c.sent();
                        return [4, this.createLoginState()];
                    case 3:
                        loginState = _c.sent();
                        return [4, this.getSession()];
                    case 4:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 5:
                        error_29 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_29) }];
                    case 6: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithOpenId = function (_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.useWxCloud, useWxCloud = _c === void 0 ? true : _c;
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, mainFunc, loginState, _d, _e, session, error_30;
            var _this = this;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        if (!utilities_2.adapterForWxMp.isMatch()) {
                            throw Error('wx api undefined');
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        mainFunc = function (code) { return __awaiter(_this, void 0, void 0, function () {
                            var result, credentials, error_31;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        result = undefined;
                                        credentials = undefined;
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 4, , 5]);
                                        return [4, this.oauthInstance.authApi.grantProviderToken({
                                                provider_id: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                provider_code: code,
                                                provider_params: {
                                                    provider_code_type: 'open_id',
                                                    appid: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                },
                                            }, useWxCloud)];
                                    case 2:
                                        result = _a.sent();
                                        if ((result === null || result === void 0 ? void 0 : result.error_code) || !result.provider_token) {
                                            throw result;
                                        }
                                        return [4, this.oauthInstance.authApi.signInWithProvider({ provider_token: result.provider_token }, useWxCloud)];
                                    case 3:
                                        credentials = _a.sent();
                                        if (credentials === null || credentials === void 0 ? void 0 : credentials.error_code) {
                                            throw credentials;
                                        }
                                        return [3, 5];
                                    case 4:
                                        error_31 = _a.sent();
                                        throw error_31;
                                    case 5: return [4, this.oauthInstance.oauth2client.setCredentials(credentials)];
                                    case 6:
                                        _a.sent();
                                        return [2];
                                }
                            });
                        }); };
                        _f.label = 1;
                    case 1:
                        _f.trys.push([1, 5, , 6]);
                        return [4, new Promise(function (resolve, reject) {
                                wx.login({
                                    success: function (res) { return __awaiter(_this, void 0, void 0, function () {
                                        var error_32;
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    _a.trys.push([0, 2, , 3]);
                                                    return [4, mainFunc(res.code)];
                                                case 1:
                                                    _a.sent();
                                                    resolve(true);
                                                    return [3, 3];
                                                case 2:
                                                    error_32 = _a.sent();
                                                    reject(error_32);
                                                    return [3, 3];
                                                case 3: return [2];
                                            }
                                        });
                                    }); },
                                    fail: function (res) {
                                        var error = new Error(res === null || res === void 0 ? void 0 : res.errMsg);
                                        error.code = res === null || res === void 0 ? void 0 : res.errno;
                                        reject(error);
                                    },
                                });
                            })];
                    case 2:
                        _f.sent();
                        return [4, this.createLoginState()];
                    case 3:
                        loginState = _f.sent();
                        return [4, this.getSession()];
                    case 4:
                        _d = (_f.sent()).data, _e = _d === void 0 ? {} : _d, session = _e.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 5:
                        error_30 = _f.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_30) }];
                    case 6: return [2];
                }
            });
        });
    };
    SbaseApi.prototype.signInWithPhoneAuth = function (_a) {
        var _b = _a.phoneCode, phoneCode = _b === void 0 ? '' : _b;
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, providerInfo, code, providerToken, signInRes, error_33, loginState, _c, _d, session;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (!utilities_2.adapterForWxMp.isMatch()) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'wx api undefined' }) }];
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        providerInfo = {
                            provider_params: { provider_code_type: 'phone' },
                            provider_id: wxInfo.appId,
                        };
                        return [4, wx.login()];
                    case 1:
                        code = (_e.sent()).code;
                        providerInfo.provider_code = code;
                        _e.label = 2;
                    case 2:
                        _e.trys.push([2, 6, , 7]);
                        return [4, this.oauthInstance.authApi.grantProviderToken(providerInfo)];
                    case 3:
                        providerToken = _e.sent();
                        if (providerToken.error_code) {
                            throw providerToken;
                        }
                        return [4, this.oauthInstance.authApi.patchProviderToken({
                                provider_token: providerToken.provider_token,
                                provider_id: wxInfo.appId,
                                provider_params: {
                                    code: phoneCode,
                                    provider_code_type: 'phone',
                                },
                            })];
                    case 4:
                        providerToken = _e.sent();
                        if (providerToken.error_code) {
                            throw providerToken;
                        }
                        return [4, this.oauthInstance.authApi.signInWithProvider({
                                provider_token: providerToken.provider_token,
                            })];
                    case 5:
                        signInRes = _e.sent();
                        if (signInRes === null || signInRes === void 0 ? void 0 : signInRes.error_code) {
                            throw signInRes;
                        }
                        return [3, 7];
                    case 6:
                        error_33 = _e.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_33) }];
                    case 7: return [4, this.createLoginState()];
                    case 8:
                        loginState = _e.sent();
                        return [4, this.getSession()];
                    case 9:
                        _c = (_e.sent()).data, _d = _c === void 0 ? {} : _c, session = _d.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                }
            });
        });
    };
    SbaseApi.prototype.formatPhone = function (phone) {
        if (!/\s+/.test(phone) && /^\+\d{1,3}\d+/.test(phone)) {
            return phone.replace(/^(\+\d{1,2})(\d+)$/, '$1 $2');
        }
        return /^\+\d{1,3}\s+/.test(phone) ? phone : "+86 ".concat(phone);
    };
    SbaseApi.prototype.notifyListeners = function (event, session, info) {
        this.listeners.forEach(function (callbacks) {
            callbacks.forEach(function (callback) {
                try {
                    callback(event, session, info);
                }
                catch (error) {
                    console.error('Error in auth state change callback:', error);
                }
            });
        });
        return;
    };
    SbaseApi.prototype.init = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var credentials, error_34;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4, this.oauthInstance.oauth2client.localCredentials.getCredentials()];
                    case 1:
                        credentials = _b.sent();
                        if (credentials) {
                            (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.INITIAL_SESSION });
                        }
                        return [3, 3];
                    case 2:
                        error_34 = _b.sent();
                        return [3, 3];
                    case 3: return [2, { error: null }];
                }
            });
        });
    };
    SbaseApi.prototype.setupListeners = function () {
        var _this = this;
        var _a;
        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.on(oauth_1.EVENTS.AUTH_STATE_CHANGED, function (params) { return __awaiter(_this, void 0, void 0, function () {
            var event, info, session;
            var _a, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        event = (_a = params === null || params === void 0 ? void 0 : params.data) === null || _a === void 0 ? void 0 : _a.event;
                        info = (_b = params === null || params === void 0 ? void 0 : params.data) === null || _b === void 0 ? void 0 : _b.info;
                        return [4, this.getSession()];
                    case 1:
                        session = (_c.sent()).data.session;
                        this.notifyListeners(event, session, info);
                        return [2];
                }
            });
        }); });
    };
    SbaseApi.prototype.convertToUser = function (userInfo) {
        var _a, _b, _c;
        if (!userInfo)
            return null;
        var email = (userInfo === null || userInfo === void 0 ? void 0 : userInfo.email) || (userInfo === null || userInfo === void 0 ? void 0 : userInfo.username);
        var phone = userInfo === null || userInfo === void 0 ? void 0 : userInfo.phone_number;
        var userId = (userInfo === null || userInfo === void 0 ? void 0 : userInfo.sub) || (userInfo === null || userInfo === void 0 ? void 0 : userInfo.uid) || '';
        return {
            id: userId,
            aud: 'authenticated',
            role: userInfo.groups,
            email: email || '',
            email_confirmed_at: (userInfo === null || userInfo === void 0 ? void 0 : userInfo.email_verified) ? userInfo.created_at : userInfo.created_at,
            phone: phone,
            phone_confirmed_at: phone ? userInfo.created_at : undefined,
            confirmed_at: userInfo.created_at,
            last_sign_in_at: userInfo.last_sign_in_at,
            app_metadata: {
                provider: ((_a = userInfo.loginType) === null || _a === void 0 ? void 0 : _a.toLowerCase()) || 'cloudbase',
                providers: [((_b = userInfo.loginType) === null || _b === void 0 ? void 0 : _b.toLowerCase()) || 'cloudbase'],
            },
            user_metadata: {
                name: userInfo === null || userInfo === void 0 ? void 0 : userInfo.name,
                picture: userInfo === null || userInfo === void 0 ? void 0 : userInfo.picture,
                username: userInfo === null || userInfo === void 0 ? void 0 : userInfo.username,
                gender: userInfo === null || userInfo === void 0 ? void 0 : userInfo.gender,
                locale: userInfo === null || userInfo === void 0 ? void 0 : userInfo.locale,
                uid: userInfo.uid,
                nickName: userInfo.nickName,
                avatarUrl: userInfo.avatarUrl || userInfo.picture,
                location: userInfo.location,
                hasPassword: userInfo.hasPassword,
            },
            identities: ((_c = userInfo === null || userInfo === void 0 ? void 0 : userInfo.providers) === null || _c === void 0 ? void 0 : _c.map(function (p) { return ({
                id: p.id || '',
                identity_id: p.id || '',
                user_id: userId,
                identity_data: {
                    provider_id: p.id,
                    provider_user_id: p.provider_user_id,
                    name: p.name,
                },
                provider: p.id || 'cloudbase',
                created_at: userInfo.created_at,
                updated_at: userInfo.updated_at,
                last_sign_in_at: userInfo.last_sign_in_at,
            }); })) || [],
            created_at: userInfo.created_at,
            updated_at: userInfo.updated_at,
            is_anonymous: userInfo.name === 'anonymous',
        };
    };
    SbaseApi.prototype.validateParams = function (params, rules) {
        for (var _i = 0, _a = Object.entries(rules); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], rule = _b[1];
            if (rule.required && ((params === null || params === void 0 ? void 0 : params[key]) === undefined || (params === null || params === void 0 ? void 0 : params[key]) === null || (params === null || params === void 0 ? void 0 : params[key]) === '')) {
                throw new oauth_1.AuthError({ message: rule.message });
            }
        }
    };
    SbaseApi.prototype.validateAtLeastOne = function (params, fieldGroups, message) {
        var hasValue = fieldGroups.some(function (group) { return group.some(function (field) { return (params === null || params === void 0 ? void 0 : params[field]) !== undefined && (params === null || params === void 0 ? void 0 : params[field]) !== null && (params === null || params === void 0 ? void 0 : params[field]) !== ''; }); });
        if (!hasValue) {
            throw new oauth_1.AuthError({ message: message });
        }
    };
    return SbaseApi;
}());
exports.SbaseApi = SbaseApi;
