// @tcwd/weapps-sdk/lib/app-h5-sdk/interactive/actionSheet.js
import { __assign } from "tslib";
import { inlineStyle, setTransform, calPxToREM } from '../utils';
var noop = function () { };
var ActionSheet = /** @class */ (function () {
    function ActionSheet() {
        this.options = {
            itemList: [],
            itemColor: '#000000',
            success: noop,
            fail: noop,
            complete: noop
        };
        this.style = {
            maskStyle: {
                position: 'fixed',
                'z-index': '1000',
                top: '0',
                right: '0',
                left: '0',
                bottom: '0',
                background: 'rgba(0,0,0,0.6)'
            },
            actionSheetStyle: {
                'z-index': '4999',
                position: 'fixed',
                left: '0',
                bottom: '0',
                '-webkit-transform': 'translate(0, 100%)',
                transform: 'translate(0, 100%)',
                width: '100%',
                'line-height': '1.6',
                background: '#EFEFF4',
                '-webkit-transition': '-webkit-transform .3s',
                transition: 'transform .3s'
            },
            menuStyle: {
                'background-color': '#FCFCFD'
            },
            cellStyle: {
                position: 'relative',
                padding: calPxToREM(10) + " 0",
                'text-align': 'center',
                'font-size': calPxToREM(18)
            },
            cancelStyle: {
                'margin-top': calPxToREM(6),
                padding: calPxToREM(10) + " 0",
                'text-align': 'center',
                'font-size': calPxToREM(18),
                color: '#000000',
                'background-color': '#FCFCFD'
            }
        };
        this.lastConfig = {};
        this.document = document;
    }
    ActionSheet.prototype.create = function (options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        // style
        var _a = this.style, maskStyle = _a.maskStyle, actionSheetStyle = _a.actionSheetStyle, menuStyle = _a.menuStyle, cellStyle = _a.cellStyle, cancelStyle = _a.cancelStyle;
        // configuration
        var config = __assign(__assign({}, this.options), options);
        this.lastConfig = config;
        // wrapper
        this.el = config.document.createElement('div');
        this.el.className = 'weapps__actionSheet';
        this.el.style.opacity = '0';
        this.el.style.transition = 'opacity 0.2s linear';
        // mask
        var mask = config.document.createElement('div');
        mask.setAttribute('style', inlineStyle(maskStyle));
        // actionSheet
        this.actionSheet = config.document.createElement('div');
        this.actionSheet.setAttribute('style', inlineStyle(actionSheetStyle));
        // menu
        this.menu = config.document.createElement('div');
        this.menu.setAttribute('style', inlineStyle(__assign(__assign({}, menuStyle), { color: config.itemColor })));
        // cells
        this.cells = config.itemList.map(function (item, index) {
            var cell = config.document.createElement('div');
            cell.className = 'weapps-actionsheet__cell';
            cell.setAttribute('style', inlineStyle(cellStyle));
            cell.textContent = item;
            cell.dataset.tapIndex = index;
            cell.onclick = function (e) { return _this.onCellClick(e); };
            return cell;
        });
        // cancel
        this.cancel = config.document.createElement('div');
        this.cancel.setAttribute('style', inlineStyle(cancelStyle));
        this.cancel.textContent = '取消';
        // result
        this.cells.forEach(function (item) { return _this.menu.appendChild(item); });
        this.actionSheet.appendChild(this.menu);
        this.actionSheet.appendChild(this.cancel);
        this.el.appendChild(mask);
        this.el.appendChild(this.actionSheet);
        // callbacks
        var cb = function () {
            _this.hide();
            var res = { errMsg: 'showActionSheet:fail cancel' };
            config.fail(res);
            config.complete(res);
            _this.rejectHandler(res);
        };
        mask.onclick = cb;
        this.cancel.onclick = cb;
        // show immediately
        config.document.body.appendChild(this.el);
        setTimeout(function () {
            _this.el.style.opacity = '1';
            setTransform(_this.actionSheet, 'translate(0, 0)');
        }, 0);
        return new Promise(function (resolve, reject) {
            _this.resolveHandler = resolve;
            _this.rejectHandler = reject;
        });
    };
    ActionSheet.prototype.show = function (options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var config = __assign(__assign({}, this.options), options);
        this.lastConfig = config;
        if (this.hideOpacityTimer)
            clearTimeout(this.hideOpacityTimer);
        if (this.hideDisplayTimer)
            clearTimeout(this.hideDisplayTimer);
        // itemColor
        if (config.itemColor)
            this.menu.style.color = config.itemColor;
        // cells
        var cellStyle = this.style.cellStyle;
        config.itemList.forEach(function (item, index) {
            var cell;
            if (_this.cells[index]) {
                // assign new content
                cell = _this.cells[index];
            }
            else {
                // create new cell
                cell = config.document.createElement('div');
                cell.className = 'weapps-actionsheet__cell';
                cell.setAttribute('style', inlineStyle(cellStyle));
                cell.dataset.tapIndex = index;
                _this.cells.push(cell);
                _this.menu.appendChild(cell);
            }
            cell.textContent = item;
            cell.onclick = function (e) { return _this.onCellClick(e); };
        });
        var cellsLen = this.cells.length;
        var itemListLen = config.itemList.length;
        if (cellsLen > itemListLen) {
            for (var i = itemListLen; i < cellsLen; i++) {
                this.menu.removeChild(this.cells[i]);
            }
            this.cells.splice(itemListLen);
        }
        // show
        this.el.style.display = 'block';
        setTimeout(function () {
            _this.el.style.opacity = '1';
            setTransform(_this.actionSheet, 'translate(0, 0)');
        }, 0);
        return new Promise(function (resolve, reject) {
            _this.resolveHandler = resolve;
            _this.rejectHandler = reject;
        });
    };
    ActionSheet.prototype.onCellClick = function (e) {
        this.hide();
        var res = {
            errMsg: 'showActionSheet:ok',
            tapIndex: +e.currentTarget.dataset.tapIndex
        };
        this.lastConfig.success && this.lastConfig.success(res);
        this.lastConfig.complete && this.lastConfig.complete(res);
        this.resolveHandler(res);
    };
    ActionSheet.prototype.hide = function () {
        var _this = this;
        if (this.hideOpacityTimer)
            clearTimeout(this.hideOpacityTimer);
        if (this.hideDisplayTimer)
            clearTimeout(this.hideDisplayTimer);
        this.hideOpacityTimer = setTimeout(function () {
            _this.el.style.opacity = '0';
            setTransform(_this.actionSheet, 'translate(0, 100%)');
            _this.hideDisplayTimer = setTimeout(function () { _this.el.style.display = 'none'; }, 200);
        }, 0);
    };
    return ActionSheet;
}());
export default ActionSheet;
