// @tcwd/weapps-sdk/lib/app-h5-sdk/interactive/modal.js
import { __assign } from "tslib";
import { inlineStyle, calPxToREM } from '../utils';
var noop = function () { };
var Modal = /** @class */ (function () {
    function Modal() {
        this.options = {
            title: '',
            content: '',
            showCancel: true,
            cancelText: '取消',
            cancelColor: '#000000',
            confirmText: '确定',
            confirmColor: '#3CC51F',
            success: noop,
            fail: noop,
            complete: noop,
            document: document
        };
        this.style = {
            maskStyle: {
                position: 'fixed',
                'z-index': '1000',
                top: '0',
                right: '0',
                left: '0',
                bottom: '0',
                background: 'rgba(0,0,0,0.6)'
            },
            modalStyle: {
                'z-index': '4999',
                position: 'fixed',
                top: '50%',
                left: '50%',
                transform: 'translate(-50%, -50%)',
                width: '80%',
                'max-width': calPxToREM(300),
                'border-radius': calPxToREM(10),
                'text-align': 'center',
                'line-height': '1.6',
                overflow: 'hidden',
                background: '#FFFFFF'
            },
            titleStyle: {
                padding: calPxToREM(20) + " " + calPxToREM(24) + " " + calPxToREM(9),
                'font-size': calPxToREM(18)
            },
            textStyle: {
                padding: "0 " + calPxToREM(24) + " " + calPxToREM(12),
                'min-height': calPxToREM(40),
                'font-size': calPxToREM(15),
                'line-height': '1.3',
                color: '#808080'
            },
            footStyle: {
                position: 'relative',
                'line-height': calPxToREM(48),
                'font-size': calPxToREM(18),
                display: 'flex'
            },
            btnStyle: {
                position: 'relative',
                '-webkit-box-flex': '1',
                '-webkit-flex': '1',
                flex: '1',
                'font-size': calPxToREM(16)
            }
        };
    }
    Modal.prototype.create = function (options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        // style
        var _a = this.style, maskStyle = _a.maskStyle, modalStyle = _a.modalStyle, titleStyle = _a.titleStyle, textStyle = _a.textStyle, footStyle = _a.footStyle, btnStyle = _a.btnStyle;
        // configuration
        var config = __assign(__assign({}, this.options), options);
        // wrapper
        this.el = config.document.createElement('div');
        this.el.className = 'weapps__modal';
        this.el.style.position = 'relative';
        this.el.style.opacity = '0';
        this.el.style.transition = 'opacity 0.2s linear';
        this.el.style.zIndex = '999';
        // mask
        var mask = config.document.createElement('div');
        mask.setAttribute('style', inlineStyle(maskStyle));
        // modal
        var modal = config.document.createElement('div');
        modal.setAttribute('style', inlineStyle(modalStyle));
        // title
        var titleCSS = config.title ? __assign(__assign({}, titleStyle), { color: config.titleColor || titleStyle.color }) : __assign(__assign({}, titleStyle), { display: 'none' });
        this.title = config.document.createElement('div');
        this.title.setAttribute('style', inlineStyle(titleCSS));
        this.title.textContent = config.title;
        // text
        var textCSS = config.title ? __assign(__assign(__assign({}, config.textStyle || {}), textStyle), { color: config.contentColor || textStyle.color }) : __assign(__assign(__assign({}, config.textStyle || {}), textStyle), { padding: calPxToREM(40) + " " + calPxToREM(20) + " " + calPxToREM(26), color: config.contentColor || '#353535' });
        this.text = config.document.createElement('div');
        this.text.setAttribute('style', inlineStyle(textCSS));
        this.text.textContent = config.content;
        // foot
        var foot = config.document.createElement('div');
        foot.className = 'weapps-modal__foot';
        foot.setAttribute('style', inlineStyle(footStyle));
        // cancel button
        var cancelCSS = __assign(__assign({}, btnStyle), { color: config.cancelColor, display: config.showCancel ? 'block' : 'none' });
        this.cancel = config.document.createElement('div');
        this.cancel.className = 'weapps-model__btn';
        this.cancel.setAttribute('style', inlineStyle(cancelCSS));
        this.cancel.textContent = config.cancelText;
        this.cancel.onclick = function () {
            _this.hide();
            var res = _this.getRes('cancel');
            config.success(res);
            config.complete(res);
            _this.resolveHandler(res);
        };
        // confirm button
        this.confirm = config.document.createElement('div');
        this.confirm.className = 'weapps-model__btn';
        this.confirm.setAttribute('style', inlineStyle(btnStyle));
        this.confirm.style.color = config.confirmColor;
        this.confirm.textContent = config.confirmText;
        this.confirm.onclick = function () {
            _this.hide();
            var res = _this.getRes('confirm');
            config.success(res);
            config.complete(res);
            _this.resolveHandler(res);
        };
        // result
        foot.appendChild(this.cancel);
        foot.appendChild(this.confirm);
        modal.appendChild(this.title);
        modal.appendChild(this.text);
        modal.appendChild(foot);
        this.el.appendChild(mask);
        this.el.appendChild(modal);
        // show immediately
        config.document.body.appendChild(this.el);
        setTimeout(function () { _this.el.style.opacity = '1'; }, 0);
        return new Promise(function (resolve) { return (_this.resolveHandler = resolve); });
    };
    Modal.prototype.getRes = function (type) {
        var res = {
            errMsg: 'showModal:ok',
            cancel: false,
            confirm: false
        };
        res[type] = true;
        return res;
    };
    Modal.prototype.show = function (options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var config = __assign(__assign({}, this.options), options);
        if (this.hideOpacityTimer)
            clearTimeout(this.hideOpacityTimer);
        if (this.hideDisplayTimer)
            clearTimeout(this.hideDisplayTimer);
        // title & text
        var _a = this.style, textStyle = _a.textStyle, titleStyle = _a.titleStyle;
        var textCSS;
        if (config.title) {
            this.title.textContent = config.title;
            // none => block
            this.title.style.display = 'block';
            var titleCSS = __assign(__assign({}, titleStyle), { color: config.titleColor || titleStyle.color || '' });
            this.title.setAttribute('style', inlineStyle(titleCSS));
            var textCss = __assign(__assign({}, textStyle), { color: config.contentColor || titleStyle.color || '' });
            this.text.setAttribute('style', inlineStyle(textCss));
        }
        else {
            // block => none
            this.title.style.display = 'none';
            textCSS = __assign(__assign({}, textStyle), { padding: calPxToREM(40) + " " + calPxToREM(20) + " " + calPxToREM(26), color: config.contentColor || '#353535' });
            this.text.setAttribute('style', inlineStyle(textCSS));
        }
        this.text.textContent = config.content || '';
        // showCancel
        this.cancel.style.display = config.showCancel ? 'block' : 'none';
        // cancelText
        this.cancel.textContent = config.cancelText || '';
        // cancelColor
        this.cancel.style.color = config.cancelColor || undefined;
        // confirmText
        this.confirm.textContent = config.confirmText || '';
        // confirmColor
        this.confirm.style.color = config.confirmColor || undefined;
        // cbs
        this.cancel.onclick = function () {
            _this.hide();
            var res = _this.getRes('cancel');
            config.success(res);
            config.complete(res);
            _this.resolveHandler(res);
        };
        this.confirm.onclick = function () {
            _this.hide();
            var res = _this.getRes('confirm');
            config.success(res);
            config.complete(res);
            _this.resolveHandler(res);
        };
        // show
        this.el.style.display = 'block';
        setTimeout(function () { _this.el.style.opacity = '1'; }, 0);
        return new Promise(function (resolve) { return (_this.resolveHandler = resolve); });
    };
    Modal.prototype.hide = function () {
        var _this = this;
        if (this.hideOpacityTimer)
            clearTimeout(this.hideOpacityTimer);
        if (this.hideDisplayTimer)
            clearTimeout(this.hideDisplayTimer);
        this.hideOpacityTimer = setTimeout(function () {
            _this.el.style.opacity = '0';
            _this.hideDisplayTimer = setTimeout(function () { _this.el.style.display = 'none'; }, 200);
        }, 0);
    };
    return Modal;
}());
export default Modal;
