// @cloudbase/weda-client/dist/cjs/mp/runtime/flow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateEventFlows = exports.EventFlow = void 0;
const util_1 = require("./util");
const event_emitter_1 = require("./event-emitter");
class EventFlow {
    id;
    description;
    #eventHandlerMap = {};
    #context = {};
    constructor({ schema, context, options }) {
        this.#eventHandlerMap = {};
        this.#context = context;
        const { id, description, eventHandlers = {} } = schema || {};
        this.id = id || '';
        this.description = description || '';
        this.#eventHandlerMap = Object.entries((0, util_1.createEventHandlers)(eventHandlers, {
            looseError: true,
            isComposite: options?.isComposite || false,
            syncCall: true,
        })).reduce((map, [key, fn]) => {
            map[key] = fn;
            return map;
        }, {
            _getInstance: () => this.#context.$page || this.#context.$app,
        });
        return this;
    }
    trigger(additionalScope, options = {}) {
        const mergedContext = {
            ...this.#context,
            ...options,
        };
        const { target } = mergedContext;
        const eventName = `${this.id}.start`;
        return this.#eventHandlerMap[(0, util_1.getMpEventHandlerName)(this.id, eventName)](new event_emitter_1.Event({
            type: eventName,
            detail: additionalScope,
            target,
            currentTarget: target,
        }));
    }
}
exports.EventFlow = EventFlow;
function generateEventFlows(flows = [], context, options) {
    const result = {};
    for (let flow of flows) {
        result[flow.id] = new EventFlow({ schema: flow, context, options });
    }
    return result;
}
exports.generateEventFlows = generateEventFlows;
