// @cloudbase/weda-client/dist/cjs/mp/runtime/weapp-component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createComponent = void 0;
const mobx_1 = require("mobx");
const util_1 = require("./util");
const query_1 = require("./query");
const flow_1 = require("./flow");
const widget_1 = require("./widget");
const merge_renderer_1 = __importDefault(require("./merge-renderer"));
const watch_1 = require("./watch");
const vender_1 = require("../../vender");
const widget_2 = require("./widget");
const weda_cloud_sdk_1 = require("@cloudbase/weda-cloud-sdk");
const event_emitter_1 = require("./event-emitter");
const behaviors_1 = require("../behaviors");
const app_1 = require("../app");
function createComponent({ key, behaviors, properties = {}, events, handler, dataBinds = {}, evtListeners, widgetProps, lifeCycle, stateFn, computedFuncs, config, query: datasetQuery = {}, eventFlows = [], datasetProfile = undefined, nativeMode = false, }) {
    /**
     * 单次引用
     */
    if (datasetProfile) {
        (0, weda_cloud_sdk_1.setDatasetProfiles)({ [key]: datasetProfile });
    }
    widgetProps = (0, util_1.patchWdigetPropsWithEvtListeners)(widgetProps, evtListeners);
    util_1.compLowcodes[key] = {
        index: {},
        stateFn,
        computedFuncs,
        handler,
        // events,
        lib: { const: {}, tools: {} },
        config,
    };
    return Component({
        options: {
            virtualHost: true,
            multipleSlots: true,
            styleIsolation: 'shared',
        },
        /**
         * commonCompBehavior 生命周期最先执行
         */
        behaviors: [behaviors_1.commonCompBehavior, ...behaviors],
        // externalClasses: ['class'],
        properties: {
            style: {
                type: String,
            },
            className: {
                type: String,
            },
            ...properties,
        },
        data: (0, widget_2.createInitData)(widgetProps, dataBinds, ''),
        lifetimes: {
            created() {
                this._pageActive = true;
                this._disposers = [];
                this._nativeObserver = false;
                this._nativeMode = nativeMode;
            },
            attached() {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                $comp.__internal__.active = this._pageActive;
                $comp.props.events = createPropEvents(events, this);
                const { widgets, rootWidget: virtualRootWidget } = (0, widget_1.createWidgets)(widgetProps, dataBinds, this, $comp.widgets);
                $comp.widgets = widgets;
                this._virtualRootWidget = virtualRootWidget;
                for (const queryId in $comp.dataset?.query || {}) {
                    if ($comp.dataset.query[queryId]?._schema?.trigger === 'auto') {
                        /**
                         * query 初始化不阻塞生命周期
                         */
                        Promise.resolve()
                            .then(async () => {
                            await $comp.dataset.query[queryId]?.initPromise;
                            return $comp.dataset.query[queryId].trigger();
                        })
                            .catch((e) => {
                            console.error(`query ${queryId} 初始化失败：`, e);
                        });
                    }
                }
                try {
                    lifeCycle.onAttached && lifeCycle.onAttached.call($comp);
                    this._invokeEventHandler('attached');
                    this.__mnt__ = (e) => {
                        const widget = (0, widget_1.getWidget)($comp.widgets, e.target.id);
                        widget._methods = e.detail.methods;
                    };
                    this.__unmnt__ = (e) => {
                        const widget = (0, vender_1.get)($comp.widgets, e.target.id);
                        widget._methods = {};
                    };
                    if ($comp.methods) {
                        this.triggerEvent('attached', {
                            methods: $comp.methods,
                        });
                    }
                }
                catch (e) {
                    console.error('Component lifecycle(attached) error', this.is, e);
                }
                this._disposers.push(...this.initMergeRenderer($comp.widgets));
            },
            ready() {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                this._disposers.push(...(0, watch_1.runWatchers)({}, this));
                lifeCycle.onReady && lifeCycle.onReady.call($comp);
                this._invokeEventHandler('ready');
            },
            detached() {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                this._pageActive = false;
                $comp.__internal__.active = this._pageActive;
                $comp.widgets = null;
                $comp.node._eventListeners?.clear?.();
                (0, widget_1.disposeWidget)(this._virtualRootWidget);
                this._disposers.forEach((dispose) => dispose());
                lifeCycle.onDetached && lifeCycle.onDetached.call($comp);
                this._invokeEventHandler('detached');
                if ($comp?.methods) {
                    this.triggerEvent('detached');
                }
            },
        },
        pageLifetimes: {
            show() {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                lifeCycle.onPageShow && lifeCycle.onPageShow.call($comp);
                this._invokeEventHandler('show');
            },
            hide() {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                lifeCycle.onPageHide && lifeCycle.onPageHide.call($comp);
                this._invokeEventHandler('hide');
            },
            // @ts-ignore
            resize(size) {
                const $comp = this._getInstance();
                if (!$comp)
                    return;
                lifeCycle.onPageResize && lifeCycle.onPageResize.call($comp, size);
            },
        },
        methods: {
            ...(0, util_1.createEventHandlers)(evtListeners, { looseError: true, isComposite: true }),
            ...merge_renderer_1.default,
            _invokeEventHandler(triggerName, params = {}) {
                const keyName = (0, util_1.getMpEventHandlerName)('__comp__', triggerName);
                const event = new event_emitter_1.Event({
                    type: triggerName,
                    detail: params,
                });
                return this[keyName]?.call?.(this, event);
            },
            _getInstance() {
                if (!this.$WEAPPS_COMP) {
                    let widget = this.$node;
                    if (!widget) {
                        if ((this.selectOwnerComponent && !this.selectOwnerComponent?.()) || this._nativeMode) {
                            const widgetValue = {};
                            for (const key in properties) {
                                if (properties[key]?.value !== undefined) {
                                    widgetValue[key] = properties[key]?.value;
                                }
                                if (this.data[key] !== undefined) {
                                    widgetValue[key] = this.data[key];
                                }
                            }
                            this.$node = (0, widget_2.createWidget)({ widgetType: key, ...widgetValue }, this.id || `id${Date.now()}`, '', undefined, {}, {});
                            widget = this.$node;
                            this._nativeObserver = true;
                        }
                        else {
                            console.error('Fatal error: weapps component instance not created', this.is, this.id);
                            return;
                        }
                    }
                    widget.getDom = (fields) => this._virtualRootWidget.children[0].getDom(fields);
                    const { $comp, disposes = [] } = create$comp(widget, { datasetQuery, eventFlows });
                    this.$WEAPPS_COMP = $comp;
                    this._disposers.push(...disposes);
                }
                return this.$WEAPPS_COMP;
            },
        },
        observers: createObservers(Object.keys(properties)),
    });
}
exports.createComponent = createComponent;
// The component instance for lowcode
function create$comp(w, { eventFlows, datasetQuery }) {
    const lowcode = util_1.compLowcodes[w.widgetType];
    if (!lowcode) {
        return {};
    }
    const { stateFn, computedFuncs, handler, lib } = lowcode;
    const { $w, app } = (0, app_1.getWedaAPI)() || {};
    const context = {
        $w: new Proxy($w || {}, {
            get(target, prop) {
                if (prop === '$comp') {
                    return $comp;
                }
                if (prop === '$page' || prop === 'page') {
                    return (0, app_1.getWedaAPI)()?.app?.__internal__.activePage;
                }
                // 代理 query
                if ($comp.dataset?.query?.[prop]) {
                    return $comp.dataset.query[prop];
                }
                // 代理 flow
                if ($comp.__internal__.eventFlows?.[prop]) {
                    return $comp.__internal__.eventFlows[prop];
                }
                // 尝试代理组件级别组件实例
                const childWidget = $comp.widgets?.[prop];
                if (childWidget) {
                    return childWidget._userWidget;
                }
                return target[prop];
            },
        }),
        $app: app,
        $page: new Proxy({}, {
            get(target, p) {
                return (0, app_1.getWedaAPI)()?.app.utils.getCurrentPage()?.[p];
            },
        }),
    };
    const $comp = {
        __internal__: {
            $w: context.$w,
            eventFlows: (0, flow_1.generateEventFlows)(eventFlows, context),
        },
        state: {},
        computed: {},
        widgets: {},
        node: w || {},
        props: {
            data: w || {},
            events: {},
            get style() {
                return w.style;
            },
            get classList() {
                return w.classList;
            },
        },
        handler: {},
        lib,
    };
    context.$comp = $comp;
    $comp.$WEAPPS_COMP = $comp; // TODO $comp will replaced to this.$WEAPPS_COMP
    $comp.computed = (0, util_1.createComputed)(computedFuncs, $comp);
    $comp.handler = Object.keys(handler).reduce((result, key) => {
        result[key] = handler[key].bind($comp);
        return result;
    }, {});
    $comp.state = (0, mobx_1.observable)(stateFn.call($comp)); // May depend on this.props.data.xxx
    let dataset = (0, weda_cloud_sdk_1.createDataset)(w.widgetType, undefined, {
        appId: app?.id,
    });
    (0, weda_cloud_sdk_1.createStateDataSourceVar)(w.widgetType, (0, weda_cloud_sdk_1.generateParamsParser)(context));
    dataset.query = (0, query_1.generateDatasetQuery)(datasetQuery, context);
    $comp.dataset = dataset;
    return {
        $comp,
        disposes: Object.values(dataset.query).map((query) => {
            return function dispose() {
                try {
                    query?.destroy?.();
                }
                catch (e) { }
            };
        }),
    };
}
function createObservers(props) {
    const MAP = {
        className: {
            alias: 'classList',
            format: function (value = '') {
                return Array.from(new Set(value.split(' ').filter((item) => !!item)));
            },
        },
    };
    return props.reduce((observers, prop) => {
        observers[prop] = function (value) {
            if (!this._nativeObserver) {
                return;
            }
            const $comp = this._getInstance();
            if ($comp) {
                const data = $comp.props.data || {};
                const dataKey = MAP[prop]?.alias || prop;
                const formatValue = MAP[prop]?.format ? MAP[prop].format(value) : value;
                // if (!deepEqual(data[prop], formatValue)) {
                data[dataKey] = formatValue;
                // } else {
                //   // console.log(`Same comp prop will not trigger observer. ${prop}->${dataKey}`, formatValue)
                // }
            }
        };
        return observers;
    }, {});
}
function dataBindsBindContext(dataBinds, self) {
    return Object.keys(dataBinds).reduce((result, widgetId) => {
        result[widgetId] = Object.keys(dataBinds[widgetId]).reduce((result, prop) => {
            result[prop] = dataBinds[widgetId][prop].bind(self);
            return result;
        }, {});
        return result;
    }, {});
}
function createPropEvents(events, mpInst) {
    const protectEventKeys = [
        'touchstart',
        'touchmove',
        'touchcancel',
        'touchend',
        'tap',
        'longpress',
        'longtap',
        'transitionend',
        'animationstart',
        'animationiteration',
        'animationend',
        'touchforcechange', // 在支持 3D Touch 的 iPhone 设备，重按时会触发
    ];
    const result = {};
    events.forEach((evt) => {
        const isProtectKey = protectEventKeys.some((key) => key === evt.name);
        if (isProtectKey) {
            result[evt.name] = function () { };
        }
        else {
            result[evt.name] = function (evtDetail) {
                if (evt.getValueFromEvent) {
                    mpInst.setData({ value: evt.getValueFromEvent({ detail: evtDetail }) });
                }
                mpInst.triggerEvent(evt.name, evtDetail);
                mpInst._getInstance().node?._eventListeners?.emit?.(evt.name, evtDetail);
            };
        }
    });
    return result;
}
