// @cloudbase/weda-client/dist/esm/common/calculation/declare.js
export default {
    ABS: {
        id: 'ABS',
        type: 'method',
        name: '绝对值',
        funName: 'ABS',
        gram: 'ABS(数字)',
        editorGram: 'ABS(${1:1})',
        desc: '计算传入数字的绝对值',
        cases: '计算数值-9.7 的绝对值。',
        example: 'ABS(-9.7)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 0,
    },
    Min: {
        id: 'Min',
        type: 'method',
        name: '最小值',
        funName: 'Min',
        gram: 'Min(数字1, [数字2, ...])',
        editorGram: 'Min(${1:1}, ${2:2})',
        desc: '返回一组数字中的最小值',
        cases: '获取 1，15，100，100.6 中的最下值，返回值为 1。',
        example: 'Min(1,15,100,100.6)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 1,
    },
    Max: {
        id: 'Max',
        type: 'method',
        name: '最大值',
        funName: 'Max',
        gram: 'Max(数字1, [数字2, ...])',
        editorGram: 'Max(${1:1}, ${2:2})',
        desc: '返回一组数字中的最大值',
        cases: '获取 1，15，100，100.6 中的最大值，返回值为 100.6。',
        example: 'Max(1,15,100,100.6)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 2,
    },
    Average: {
        id: 'Average',
        type: 'method',
        name: '平均值',
        funName: 'Average',
        editorGram: 'Average(${1:1}, ${2:2})',
        gram: 'Average(数字1, [数字2, ...])',
        cases: '获取 1，15，100，100.6 的平均值，返回值为 54.15。',
        example: 'Average(1,15,100,100.6)',
        desc: '返回一组数字中的平均值',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 3,
    },
    Sum: {
        id: 'Sum',
        type: 'method',
        name: '求和',
        funName: 'Sum',
        gram: 'Sum(数字1, [数字2, ...])',
        editorGram: 'Sum(${1:1}, ${2:2})',
        desc: '返回一组数字中的和',
        cases: '获取 1，15，100，100.6 的和，返回值为 216.6。',
        example: 'Sum(1,15,100,100.6)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 4,
    },
    Floor: {
        id: 'Floor',
        type: 'method',
        name: '向下取整',
        funName: 'Floor',
        gram: 'Floor(数字)',
        editorGram: 'Floor(${1:1})',
        desc: '返回传入数字向下取整的结果',
        cases: '计算数值 7.7 的下取整，返回值为 7。',
        example: 'Floor(7.7)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 5,
    },
    Ceiling: {
        id: 'Ceiling',
        type: 'method',
        name: '向上取整',
        funName: 'Ceiling',
        gram: 'Ceiling(数字)',
        editorGram: 'Ceiling(${1:1})',
        desc: '返回传入数字向上取整的结果',
        cases: '计算数值 7.1 的上取整，返回值为 8。',
        example: 'Ceiling(7.1)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 6,
    },
    Round: {
        id: 'Round',
        type: 'method',
        name: '四舍五入',
        funName: 'Round',
        gram: 'Round(数字)',
        editorGram: 'Round(${1:1})',
        desc: '返回传入数字四舍五入后的结果',
        cases: '计算 216.6 四舍五入的值，返回值为 217。',
        example: 'Round(216.6)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 7,
    },
    Rand: {
        id: 'Rand',
        type: 'method',
        name: '随机数',
        funName: 'Rand',
        gram: 'Rand(数字)',
        editorGram: 'Rand(${1:1})',
        desc: '返回一个指定范围的伪随机数，例如 Rand(10)，返回一个范围在 10 以内的随机数',
        cases: '获取一个 10 以内的随机数。',
        example: 'Rand(10)',
        schemaRetType: { type: 'number' },
        retType: 'number',
        Zindex: 8,
    },
};
