// @cloudbase/weda-client/dist/esm/web/app.js
import React from 'react';
import { observable } from 'mobx';
import { actionSdk as sdk } from './actions/action-func';
import { DS_SDK, CLOUD_SDK, EXTRA_API, getConfig } from '@cloudbase/weda-cloud-sdk';
import { EPLATFORM } from '../common/constants';
import { execOnce, createAppUtils, isMobile, createAPINamespace } from '../common/utils';
import { createWebActionsAPI } from './actions/actions';
import { createWebRouterAPI, generatePageUrl } from './router';
import { init } from './app/tcb';
import { auth } from './auth/index';
import { env } from './env/index';
import { createWebAiAPI } from './ai';
/**
 * @description 创建web下的app对象，对外暴露
 * @param initData
 * @returns
 */
export function createWebApp(initData = {}) {
    const appId = initData?.appConfig?.id || initData.id || '';
    const app = {
        __internal__: {
            $w: null,
            activePage: null,
            enumOptions: observable({}),
            getConfig: () => initData?.appConfig || {},
            resolveStaticResourceUrl(staticUrl) {
                const { _isPrivate, _deployType } = window;
                if (/^\//.test(staticUrl) && !(_isPrivate && !_deployType)) {
                    const domain = app?.__internal__?.getConfig?.()?.staticResourceDomain || app?.domain || '';
                    const url = `${location.protocol}//${domain}${staticUrl}`;
                    return url;
                }
                return staticUrl;
            },
            isMobile,
            getCloudSdkConfig: getConfig,
            generatePageUrl: (obj = {}) => {
                return generatePageUrl({ ...obj, basename: obj.basename || app?.__internal__?.getConfig?.()?.basename });
            },
        },
        id: appId,
        name: initData?.appConfig?.name || appId,
        label: initData.appConfig?.label || '',
        version: initData?.appConfig?.version || '',
        mpAppId: '',
        mpAppName: '',
        platform: EPLATFORM.WEB,
        /**
         * 须避免被wx.cloud 覆盖
         */
        cloud: CLOUD_SDK,
        dataSources: DS_SDK,
        _setStateVal(config) { },
        init,
        auth,
        invoke(params) {
            // deed to impl
        },
        ai: null, // 后续覆盖可用对象
    };
    const appWithUtils = Object.assign(app, {
        ...createWebActionsAPI(app, sdk, initData),
        utils: createAppUtils(app, initData),
    });
    const webApp = Object.assign(appWithUtils, {
        ...createWebRouterAPI(appWithUtils, sdk),
        ...createWebAiAPI(),
        /**
         * @deprecated
         */
        invoke(params) {
            const $page = webApp.utils.getCurrentPage();
            return $page.invokeComponentMethod(params);
        },
        /**
         * @deprecated
         */
        _setStateVal(config) {
            const [key, name] = config?.varPath?.split?.('.') || [];
            if (key !== '$global' && key !== '$page') {
                console.warn('[@deprecated] 禁止跨页面设置变量值，后期版本将放弃对此的支持');
            }
            if (key === '$page') {
                const $page = webApp.utils.getCurrentPage();
                $page?.setState?.({ [name]: config.val });
                return;
            }
            else {
                // @ts-ignore
                EXTRA_API.setState(config.varPath, config.val);
            }
        },
    });
    const platforms = ['WEB'];
    if (isMobile()) {
        platforms.push('MOBILE', 'MOBILEWEB');
    }
    else {
        platforms.push('PCWEB');
    }
    webApp.__internal__.$w = createAPINamespace(webApp, env, {
        device: initData?.device || {
            viewport: {
                width: window?.innerWidth || 0,
                height: window?.innerHeight || 0,
            },
            networkType: '',
        },
        platforms,
    });
    return webApp;
}
/**
 * @description 创建web下的app对象，对外暴露
 */
export const app = execOnce(createWebApp);
export const $w = new Proxy({}, {
    get(_, prop) {
        const { app: currentAppAPI, $page } = window;
        const api = currentAppAPI || app;
        const internal = currentAppAPI.__internal__ || app.__internal__;
        if (prop === 'app') {
            return api;
        }
        if (prop === 'page') {
            return internal.activePage || $page;
        }
        return internal.$w ? internal.$w?.[prop] : null;
    },
});
export const AppContext = React.createContext({});
export function getWedaAPI() {
    return {
        $app: globalThis.$app,
        app: globalThis.app,
        $w: globalThis.app?.__internal__.$w || $w,
    };
}
