// @tcwd/weapps-sdk/lib/app-h5-sdk/system/info.js
import MobileDetect from 'mobile-detect';
function getSystemInfoSync() {
    var md = new MobileDetect(navigator.userAgent);
    return {
        brand: md.mobile(),
        model: md.mobile(),
        system: md.os(),
        pixelRatio: window.devicePixelRatio,
        screenWidth: window.screen.width,
        screenHeight: window.screen.height,
        windowWidth: document.documentElement.clientWidth,
        windowHeight: document.documentElement.clientHeight,
        version: '',
        statusBarHeight: '',
        platform: navigator.platform,
        language: navigator.language,
        fontSizeSetting: '',
        SDKVersion: '' // 客户端基础库版本
    };
}
function getSystemInfo(options) {
    if (options === void 0) { options = {}; }
    var success = options.success, complete = options.complete;
    return new Promise(function (resolve) {
        var info = getSystemInfoSync();
        typeof success === 'function' && success(info);
        typeof complete === 'function' && complete(info);
        resolve(info);
    });
}
export { getSystemInfoSync, getSystemInfo };
