// @cloudbase/weda-client/dist/cjs/vender.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.split = exports.isPlainObject = exports.isObject = exports.toLower = exports.toUpper = exports.eq = exports.isNull = exports.set = exports.get = exports.Decimal = exports.dayjs = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
exports.dayjs = dayjs_1.default;
const isoWeek_1 = __importDefault(require("dayjs/plugin/isoWeek"));
const duration_1 = __importDefault(require("dayjs/plugin/duration"));
const objectSupport_1 = __importDefault(require("dayjs/plugin/objectSupport"));
dayjs_1.default.extend(duration_1.default);
dayjs_1.default.extend(isoWeek_1.default);
dayjs_1.default.extend(objectSupport_1.default);
const decimal_js_1 = __importDefault(require("decimal.js"));
exports.Decimal = decimal_js_1.default;
var lodash_es_1 = require("lodash-es");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return lodash_es_1.get; } });
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return lodash_es_1.set; } });
Object.defineProperty(exports, "isNull", { enumerable: true, get: function () { return lodash_es_1.isNull; } });
Object.defineProperty(exports, "eq", { enumerable: true, get: function () { return lodash_es_1.eq; } });
Object.defineProperty(exports, "toUpper", { enumerable: true, get: function () { return lodash_es_1.toUpper; } });
Object.defineProperty(exports, "toLower", { enumerable: true, get: function () { return lodash_es_1.toLower; } });
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return lodash_es_1.isObject; } });
Object.defineProperty(exports, "isPlainObject", { enumerable: true, get: function () { return lodash_es_1.isPlainObject; } });
Object.defineProperty(exports, "split", { enumerable: true, get: function () { return lodash_es_1.split; } });
