// @tcwd/weapps-sdk/lib/app-h5-sdk/webSocket/index.js
import SocketTask from './socketTask';
import { shouleBeObject, getParameterError } from '../utils';
var socketTasks = [];
var socketsCounter = 1;
function connectSocket(options) {
    var name = 'connectSocket';
    // options must be an Object
    var isObject = shouleBeObject(options);
    if (!isObject.res) {
        var res_1 = { errMsg: "" + name + isObject.msg };
        console.error(res_1.errMsg);
        return res_1;
    }
    var url = options.url, protocols = options.protocols, success = options.success, fail = options.fail, complete = options.complete;
    var res = { errMsg: 'connectSocket:ok' };
    // options.url must be String
    if (typeof url !== 'string') {
        res.errMsg = getParameterError({
            name: name,
            para: 'url',
            correct: 'String',
            wrong: url,
        });
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return;
    }
    // options.url must be invalid
    if (!url.startsWith('ws://') && !url.startsWith('wss://')) {
        res.errMsg = "request:fail invalid url \"" + url + "\"";
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return;
    }
    // protocols must be array
    var _protocols = Array.isArray(protocols) ? protocols : null;
    // 2 connection at most
    if (socketTasks.length > 1) {
        res.errMsg = '同时最多发起 2 个 socket 请求，更多请参考文档。';
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return;
    }
    var task = new SocketTask(url, _protocols);
    task._destroyWhenClose = function () {
        var _this = this;
        socketTasks = socketTasks.filter(function (socketTask) { return socketTask !== _this; });
    };
    socketTasks.push(task);
    res.socketTaskId = socketsCounter++;
    typeof success === 'function' && success(res);
    typeof complete === 'function' && complete(res);
    return task;
}
function onSocketOpen() {
    console.warn('Deprecated.Please use socketTask.onOpen instead.');
}
function onSocketError() {
    console.warn('Deprecated.Please use socketTask.onError instead.');
}
function sendSocketMessage() {
    console.warn('Deprecated.Please use socketTask.send instead.');
}
function onSocketMessage() {
    console.warn('Deprecated.Please use socketTask.onMessage instead.');
}
function closeSocket() {
    console.warn('Deprecated.Please use socketTask.close instead.');
}
function onSocketClose() {
    console.warn('Deprecated.Please use socketTask.onClose instead.');
}
export { connectSocket, onSocketOpen, onSocketError, sendSocketMessage, onSocketMessage, closeSocket, onSocketClose, };
