// @tcwd/weapps-sdk/lib/app-h5-sdk/webSocket/socketTask.js
var SocketTask = /** @class */ (function () {
    function SocketTask(url, protocols) {
        if (protocols && protocols.length) {
            this.ws = new WebSocket(url, protocols);
        }
        else {
            this.ws = new WebSocket(url);
        }
        this.CONNECTING = 0;
        this.OPEN = 1;
        this.CLOSING = 2;
        this.CLOSED = 3;
    }
    Object.defineProperty(SocketTask.prototype, "readyState", {
        get: function () {
            return this.ws.readyState;
        },
        enumerable: true,
        configurable: true
    });
    SocketTask.prototype.send = function (obj) {
        if (obj === void 0) { obj = {}; }
        if (typeof obj !== 'object' || !obj)
            obj = {};
        var _a = obj.data, data = _a === void 0 ? '' : _a, success = obj.success, fail = obj.fail, complete = obj.complete;
        if (this.readyState !== 1) {
            var res_1 = { errMsg: 'SocketTask.send:fail SocketTask.readState is not OPEN' };
            console.error(res_1.errMsg);
            typeof fail === 'function' && fail(res_1);
            typeof complete === 'function' && complete(res_1);
            return Promise.reject(res_1);
        }
        this.ws.send(data);
        var res = { errMsg: 'sendSocketMessage:ok' };
        typeof success === 'function' && success(res);
        typeof complete === 'function' && complete(res);
        return Promise.resolve(res);
    };
    SocketTask.prototype.close = function (obj) {
        if (obj === void 0) { obj = {}; }
        if (typeof obj !== 'object' || !obj)
            obj = {};
        var _a = obj.code, code = _a === void 0 ? 1000 : _a, _b = obj.reason, reason = _b === void 0 ? 'server complete,close' : _b, success = obj.success, complete = obj.complete;
        this.closeDetail = { code: code, reason: reason };
        // 主动断开时需要重置链接数
        this._destroyWhenClose && this._destroyWhenClose();
        this.ws.close();
        var res = { errMsg: 'closeSocket:ok' };
        typeof success === 'function' && success(res);
        typeof complete === 'function' && complete(res);
        return Promise.resolve(res);
    };
    SocketTask.prototype.onOpen = function (func) {
        this.ws.onopen = func;
    };
    SocketTask.prototype.onMessage = function (func) {
        this.ws.onmessage = func;
    };
    SocketTask.prototype.onClose = function (func) {
        var _this = this;
        this.ws.onclose = function () {
            // 若服务器方断掉也需要重置链接数
            _this._destroyWhenClose && _this._destroyWhenClose();
            func(_this.closeDetail || { code: 1006, reason: 'abnormal closure' });
        };
    };
    SocketTask.prototype.onError = function (func) {
        this.ws.onerror = func;
    };
    return SocketTask;
}());
export default SocketTask;
