// @tcwd/weapps-sdk/lib/app-h5-sdk/createSelectorQuery/index.js
import ReactDom from 'react-dom';
/**
 * WXML节点信息API
 * @return {Object} SelectorQuery 对象实例
 */
function queryBat(queue, cb) {
    var res = [];
    queue.forEach(function (item) {
        var selector = item.selector, single = item.single, fields = item.fields, component = item.component;
        // selector 的容器节点
        /* eslint-disable */
        var container = (component !== null ?
            (ReactDom.findDOMNode(component) || document) :
            document);
        /* eslint-enable */
        // 特殊处理 ---- 选自己
        var selectSelf = false;
        if (container !== document) {
            var $nodeList = container.parentNode.querySelectorAll(selector);
            for (var i = 0, len = $nodeList.length; i < len; ++i) {
                if (container === $nodeList[i]) {
                    selectSelf = true;
                    break;
                }
            }
        }
        if (single) {
            var el = selectSelf === true ? container : container.querySelector(selector);
            res.push(filter(fields, el, selector));
        }
        else {
            var $children = container.querySelectorAll(selector);
            var children = [];
            selectSelf === true && children.push(container);
            for (var i = 0, len = $children.length; i < len; ++i) {
                children.push($children[i]);
            }
            res.push(children.map(function (dom) { return filter(fields, dom); }));
        }
    });
    cb(res);
}
function filter(fields, dom, selector) {
    if (!dom)
        return null;
    var id = fields.id, dataset = fields.dataset, rect = fields.rect, size = fields.size, scrollOffset = fields.scrollOffset, _a = fields.properties, properties = _a === void 0 ? [] : _a, _b = fields.computedStyle, computedStyle = _b === void 0 ? [] : _b;
    var _c = dom.getBoundingClientRect(), left = _c.left, right = _c.right, top = _c.top, bottom = _c.bottom, width = _c.width, height = _c.height;
    var isViewport = selector === 'html';
    var res = {};
    if (id)
        res.id = dom.id;
    if (dataset)
        res.dataset = Object.assign({}, dom.dataset);
    if (rect) {
        if (!isViewport) {
            res.left = left;
            res.right = right;
            res.top = top;
            res.bottom = bottom;
        }
        else {
            res.left = 0;
            res.right = 0;
            res.top = 0;
            res.bottom = 0;
        }
    }
    if (size) {
        if (!isViewport) {
            res.width = width;
            res.height = height;
        }
        else {
            res.width = dom.clientWidth;
            res.height = dom.clientHeight;
        }
    }
    if (scrollOffset) {
        res.scrollLeft = dom.scrollLeft;
        res.scrollTop = dom.scrollTop;
        res.scrollHeight = dom.scrollHeight;
        res.scrollWidth = dom.scrollWidth;
    }
    if (properties.length) {
        properties.forEach(function (prop) {
            var attr = dom.getAttribute(prop);
            if (attr)
                res[prop] = attr;
        });
    }
    if (computedStyle.length) {
        var styles_1 = window.getComputedStyle(dom);
        computedStyle.forEach(function (key) {
            var value = styles_1.getPropertyValue(key);
            if (value)
                res[key] = value;
        });
    }
    return res;
}
var Query = /** @class */ (function () {
    function Query() {
        this._defaultWebviewId = null;
        this._webviewId = null;
        this._queue = [];
        this._queueCb = [];
        this._component = null;
    }
    Query.prototype.in = function (component) {
        this._component = component;
        return this;
    };
    Query.prototype.select = function (selector) {
        // 小程序里跨自定义组件的后代选择器 '>>>' 在 h5 替换为普通后代选择器 '>'
        if (typeof selector === 'string')
            selector = selector.replace('>>>', '>');
        return new NodesRef(selector, this, true);
    };
    Query.prototype.selectAll = function (selector) {
        // 小程序里跨自定义组件的后代选择器 '>>>' 在 h5 替换为普通后代选择器 '>'
        if (typeof selector === 'string')
            selector = selector.replace('>>>', '>');
        return new NodesRef(selector, this, false);
    };
    Query.prototype.selectViewport = function () {
        return new NodesRef('html', this, true);
    };
    Query.prototype.exec = function (cb) {
        var _this = this;
        queryBat(this._queue, function (res) {
            var _queueCb = _this._queueCb;
            res.forEach(function (item, index) {
                typeof _queueCb[index] === 'function' && _queueCb[index].call(_this, item);
            });
            typeof cb === 'function' && cb.call(_this, res);
        });
    };
    Query.prototype._push = function (selector, component, single, fields, callback) {
        if (callback === void 0) { callback = null; }
        this._queue.push({
            component: component,
            selector: selector,
            single: single,
            fields: fields
        });
        this._queueCb.push(callback);
    };
    return Query;
}());
var NodesRef = /** @class */ (function () {
    function NodesRef(selector, querySelectorQuery, single) {
        this._component = querySelectorQuery._component;
        this._selector = selector;
        this._selectorQuery = querySelectorQuery;
        this._single = single;
    }
    NodesRef.prototype.boundingClientRect = function (cb) {
        var _a = this, _selector = _a._selector, _component = _a._component, _single = _a._single, _selectorQuery = _a._selectorQuery;
        _selectorQuery._push(_selector, _component, _single, { id: !0, dataset: !0, rect: !0, size: !0 }, cb);
        return _selectorQuery;
    };
    NodesRef.prototype.scrollOffset = function (cb) {
        var _a = this, _selector = _a._selector, _component = _a._component, _single = _a._single, _selectorQuery = _a._selectorQuery;
        _selectorQuery._push(_selector, _component, _single, { id: !0, dataset: !0, scrollOffset: !0 }, cb);
        return _selectorQuery;
    };
    NodesRef.prototype.fields = function (fields, cb) {
        var _a = this, _selector = _a._selector, _component = _a._component, _single = _a._single, _selectorQuery = _a._selectorQuery;
        var id = fields.id, dataset = fields.dataset, rect = fields.rect, size = fields.size, scrollOffset = fields.scrollOffset, _b = fields.properties, properties = _b === void 0 ? [] : _b, _c = fields.computedStyle, computedStyle = _c === void 0 ? [] : _c;
        _selectorQuery._push(_selector, _component, _single, {
            id: id,
            dataset: dataset,
            rect: rect,
            size: size,
            scrollOffset: scrollOffset,
            properties: properties,
            computedStyle: computedStyle
        }, cb);
        return _selectorQuery;
    };
    return NodesRef;
}());
export function createSelectorQuery() {
    return new Query();
}
