// tcb-admin-node/src/functions/index.js
const httpRequest = require('../utils/httpRequest')

/**
 * è°ƒç”¨äº‘å‡½æ•? * @param {String} name  å‡½æ•°å? * @param {Object} functionParam å‡½æ•°å‚æ•°
 * @return {Promise}
 */
function callFunction({ name, data }) {
  try {
    data = data ? JSON.stringify(data) : ''
  } catch (e) {
    return Promise.reject(e)
  }
  if (!name) {
    return Promise.reject(
      new Error({
        message: 'å‡½æ•°åä¸èƒ½ä¸ºç©?
      })
    )
  }

  const params = {
    action: 'functions.invokeFunction',
    function_name: name,
    request_data: data
  }

  return httpRequest({
    config: this.config,
    params,
    method: 'post',
    headers: {
      'content-type': 'application/json',
      ...(process.env.TCB_ROUTE_KEY
        ? { 'X-Tcb-Route-Key': process.env.TCB_ROUTE_KEY }
        : {})
    }
  }).then(res => {
    if (res.code) {
      return res
    } else {
      let result
      try {
        result = JSON.parse(res.data.response_data)
      } catch (e) {
        result = res.data.response_data
      }
      return {
        result,
        requestId: res.requestId
      }
    }
  })
}

exports.callFunction = callFunction
