// @cloudbase/node-sdk/lib/log/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils/utils");
const code_1 = require("../const/code");
const cloudbase_1 = require("../cloudbase");
/**
 *
 *
 * @class Log
 */
class Log {
    constructor() {
        const { _SCF_TCB_LOG } = cloudbase_1.CloudBase.getCloudbaseContext();
        this.src = 'app';
        this.isSupportClsReport = true;
        if (`${_SCF_TCB_LOG}` !== '1') {
            this.isSupportClsReport = false;
        }
        else if (!console.__baseLog__) {
            this.isSupportClsReport = false;
        }
        if (!this.isSupportClsReport) {
            // å½“å‰étcb scfç¯å¢ƒ  logåŠŸèƒ½ä¼šé€€åŒ–ä¸ºconsole
            console.warn('è¯·æ£€æŸ¥æ‚¨æ˜¯å¦åœ¨æœ¬åœ°ç¯å¢?æˆ–è€?æœªå¼€é€šé«˜çº§æ—¥å¿—åŠŸèƒ½ï¼Œå½“å‰ç¯å¢ƒä¸‹æ— æ³•ä¸ŠæŠ¥clsæ—¥å¿—ï¼Œé»˜è®¤ä½¿ç”¨console');
        }
    }
    /**
     *
     *
     * @param {*} logMsg
     * @param {*} logLevel
     * @returns
     * @memberof Log
     */
    transformMsg(logMsg) {
        // ç›®å‰logMsgåªæ”¯æŒå­—ç¬¦ä¸²valueä¸”ä¸æ”¯æŒå¤šçº§, åŠ ä¸€å±‚è½¬æ¢å¤„ç?        let realMsg = {};
        realMsg = Object.assign({}, realMsg, logMsg);
        return realMsg;
    }
    /**
     *
     *
     * @param {*} logMsg
     * @param {*} logLevel
     * @memberof Log
     */
    baseLog(logMsg, logLevel) {
        // åˆ¤æ–­å½“å‰æ˜¯å¦å±äºtcb scfç¯å¢ƒ
        if (Object.prototype.toString.call(logMsg).slice(8, -1) !== 'Object') {
            throw utils_1.E(Object.assign({}, code_1.ERROR.INVALID_PARAM, { message: 'log msg must be an object' }));
        }
        const msgContent = this.transformMsg(logMsg);
        if (this.isSupportClsReport) {
            ;
            console.__baseLog__(msgContent, logLevel);
        }
        else {
            if (console[logLevel]) {
                console[logLevel](msgContent);
            }
        }
    }
    /**
     *
     *
     * @param {*} logMsg
     * @memberof Log
     */
    log(logMsg) {
        this.baseLog(logMsg, 'log');
    }
    /**
     *
     *
     * @param {*} logMsg
     * @memberof Log
     */
    info(logMsg) {
        this.baseLog(logMsg, 'info');
    }
    /**
     *
     *
     * @param {*} logMsg
     * @memberof Log
     */
    error(logMsg) {
        this.baseLog(logMsg, 'error');
    }
    /**
     *
     *
     * @param {*} logMsg
     * @memberof Log
     */
    warn(logMsg) {
        this.baseLog(logMsg, 'warn');
    }
}
exports.Log = Log;
function logger() {
    return new Log();
}
exports.logger = logger;
