// @tcwd/weapps-sdk/lib/app-h5-sdk/fileTransfer-native/common.js
import { __assign } from "tslib";
import { createCallbackManager } from '../utils';
var id = 1;
var __weappsNative = window.__weappsNative;
//DownloadTask具体属性参照小程序
var LoadTask = /** @class */ (function () {
    function LoadTask(taskApiName, identifier, json) {
        this.identifier = identifier;
        this.taskApiName = taskApiName;
        this.json = json;
        this.onProgressUpdateCb = null;
        this.onHeadersReceivedCbName = null;
        this.__callbackManager = {
            headersReceived: createCallbackManager(),
            progressUpdate: createCallbackManager(),
        };
    }
    LoadTask.prototype.abort = function () {
        var params = {
            data: {
                type: this.taskApiName,
                identifier: this.identifier,
            },
        };
        __weappsNative && __weappsNative.abort(JSON.stringify(params));
        typeof this.json.error === 'function' && this.json.error({
            errMsg: taskApiName + ":fail abort"
        });
    };
    LoadTask.prototype.onProgressUpdate = function (callback) {
        if (callback === void 0) { callback = function () { }; }
        this.__callbackManager.progressUpdate.add(callback);
        this.onProgressUpdateCb = taskApiName + this.identifier;
        if (!window[this.onProgressUpdateCb]) {
            var that_1 = this;
            window[this.onProgressUpdateCb] = function (err, args) {
                that_1.__callbackManager.progressUpdate.trigger(args ? JSON.parse(args) : {});
            };
            var params = {
                callback: this.onProgressUpdateCb,
                data: {
                    type: taskApiName,
                    identifier: this.identifier,
                },
            };
            __weappsNative && __weappsNative.onProgressUpdate(JSON.stringify(params));
        }
    };
    LoadTask.prototype.offProgressUpdate = function (callback) {
        if (callback === void 0) { callback = function () { }; }
        this.__callbackManager.progressUpdate.remove(callback);
        if (this.__callbackManager.progressUpdate.count === 0) {
            var params = {
                callback: this.onProgressUpdateCb,
                data: {
                    type: taskApiName,
                    identifier: this.identifier,
                },
            };
            __weappsNative && __weappsNative.offProgressUpdate(JSON.stringify(params));
            delete window[this.onProgressUpdateCb];
        }
    };
    LoadTask.prototype.onHeadersReceived = function (callback) {
        if (callback === void 0) { callback = function () { }; }
        this.__callbackManager.headersReceived.add(callback);
        this.onHeadersReceivedCbName = taskApiName + this.identifier;
        if (!window[this.onHeadersReceivedCbName]) {
            var that_2 = this;
            window[this.onHeadersReceivedCbName] = function (err, args) {
                that_2.__callbackManager.headersReceived.trigger(args ? JSON.parse(args) : {});
            };
            var params = {
                callback: this.onHeadersReceivedCbName,
                data: {
                    type: taskApiName,
                    identifier: this.identifier,
                },
            };
            __weappsNative && __weappsNative.onProgressUpdate(JSON.stringify(params));
        }
    };
    LoadTask.prototype.offHeadersReceived = function (callback) {
        this.__callbackManager.headersReceived.remove(callback);
        if (this.__callbackManager.headersReceived.count === 0) {
            var params = {
                callback: this.onHeadersReceivedCbName,
                data: {
                    type: taskApiName,
                    identifier: this.identifier,
                },
            };
            __weappsNative && __weappsNative.offProgressUpdate(JSON.stringify(params));
            delete window[this.onHeadersReceivedCbName];
        }
    };
    return LoadTask;
}());
export var createLoadTask = function (apiName, taskApiName, json) {
    var callbackName = apiName + id++;
    window[callbackName] = function (err, args) {
        try {
            console.log(callbackName, err, args, 'jsbridge 执行了');
            if (err !== 'null') {
                typeof json.fail === 'function' && json.fail(JSON.parse(err));
                return;
            }
            typeof json.success === 'function' && json.success(__assign({ errMsg: apiName + ":ok" }, JSON.parse(args)));
        }
        catch (e) {
            console.log(e.stack, apiName + " \u8C03\u7528\u51FA\u9519");
            typeof json.fail === 'function' && json.fail(e);
        }
        finally {
            typeof json.complete === 'function' && json.complete();
            delete window[callbackName];
        }
    };
    var data = __assign({}, json);
    delete data.success;
    delete data.fail;
    delete data.complete;
    var sdkJson = {
        data: data,
        callback: callbackName,
    };
    console.log('开始执行__weappsNative...', apiName, sdkJson, __weappsNative[apiName]);
    console.log("__weappsNative[" + apiName + "](" + JSON.stringify(sdkJson) + ")");
    var identifier = __weappsNative && __weappsNative[apiName](JSON.stringify(sdkJson));
    return new LoadTask(taskApiName, identifier, json);
};
