// @tcwd/weapps-sdk/lib/app-h5-sdk/recorderManager-native/index.js
import { __assign } from "tslib";
var __weappsNative = window.__weappsNative;
var perCallback = 'operateRecorder';
var RecorderManager = /** @class */ (function () {
    function RecorderManager() {
        this.__init__();
    }
    RecorderManager.getInstance = function () {
        if (!this.instance) {
            this.instance = new RecorderManager();
        }
        return this.instance;
    };
    RecorderManager.prototype.__triggerApi__ = function (type, object) {
        if (object === void 0) { object = {}; }
        var params = {
            data: __assign({ operationType: type }, object),
        };
        console.log("__weappsNative.operateRecorder " + type + " \u5F00\u59CB\u6267\u884C...", params);
        __weappsNative.operateRecorder(JSON.stringify(params));
    };
    RecorderManager.prototype.__triggerEvent__ = function (type, callback) {
        var callbackName = perCallback + "_" + type;
        window[callbackName] = function (err, res) {
            var resData = JSON.parse(res);
            console.log("__weappsNative.operateRecorder " + type + " event \u89E6\u53D1\u4E86...", err, resData);
            if (type === 'onFrameRecorded') {
                if (resData.frameBuffer instanceof Array) {
                    resData.frameBuffer = new Int8Array(resData.frameBuffer).buffer;
                }
            }
            typeof callback === 'function' && callback(resData);
        };
        var params = {
            callback: callbackName,
            data: {
                operationType: type,
            },
        };
        console.log("__weappsNative.operateRecorder " + type + " \u5F00\u59CB\u6267\u884C...", params);
        __weappsNative.operateRecorder(JSON.stringify(params));
    };
    RecorderManager.prototype.__init__ = function () {
        var _this = this;
        var triggerApis = ['start', 'pause', 'resume', 'stop'];
        triggerApis.forEach(function (api) {
            _this[api] = function (object) {
                if (object === void 0) { object = {}; }
                _this.__triggerApi__(api, object);
            };
        });
        var triggerEventApis = [
            'onFrameRecorded',
            'onInterruptionBegin',
            'onInterruptionEnd',
            'onPause',
            'onResume',
            'onStart',
            'onStop',
            'onError'
        ];
        triggerEventApis.forEach(function (api) {
            _this[api] = function (callback) {
                _this.__triggerEvent__(api, callback);
            };
        });
    };
    return RecorderManager;
}());
export var getRecorderManager = function () {
    if (!__weappsNative) {
        console.warn('__weappsNative doesnot exsit...');
        return null;
    }
    return RecorderManager.getInstance();
};
