// @tcwd/weapps-sdk/lib/app-h5-sdk/storage/index.js
import { shouleBeObject, getParameterError } from '../utils';
function setStorage(options) {
    // options must be an Object
    var isObject = shouleBeObject(options);
    if (!isObject.res) {
        var res_1 = { errMsg: "setStorage" + isObject.msg };
        console.error(res_1.errMsg);
        return Promise.reject(res_1);
    }
    var key = options.key, data = options.data, success = options.success, fail = options.fail, complete = options.complete;
    var res = { errMsg: 'setStorage:ok' };
    if (typeof key !== 'string') {
        res.errMsg = getParameterError({
            name: 'setStorage',
            para: 'key',
            correct: 'String',
            wrong: key
        });
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return Promise.reject(res);
    }
    setStorageSync(key, data);
    typeof success === 'function' && success(res);
    typeof complete === 'function' && complete(res);
    return Promise.resolve(res);
}
function setStorageSync(key, data) {
    if (data === void 0) { data = ''; }
    if (typeof key !== 'string') {
        console.error(getParameterError({
            name: 'setStorage',
            correct: 'String',
            wrong: key
        }));
        return;
    }
    var type = typeof data;
    var obj = {};
    if (type === 'symbol') {
        obj = { data: '' };
    }
    else {
        obj = { data: data };
    }
    localStorage.setItem(key, JSON.stringify(obj));
}
function getStorage(options) {
    // options must be an Object
    var isObject = shouleBeObject(options);
    if (!isObject.res) {
        var res_2 = { errMsg: "getStorage" + isObject.msg };
        console.error(res_2.errMsg);
        return Promise.reject(res_2);
    }
    var key = options.key, success = options.success, fail = options.fail, complete = options.complete;
    var res = { errMsg: 'getStorage:ok' };
    if (typeof key !== 'string') {
        res.errMsg = getParameterError({
            name: 'getStorage',
            para: 'key',
            correct: 'String',
            wrong: key
        });
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return Promise.reject(res);
    }
    var _a = getItem(key), result = _a.result, data = _a.data;
    if (result) {
        res.data = data;
    }
    else {
        res.errMsg = 'getStorage:fail data not found';
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return Promise.reject(res);
    }
    typeof success === 'function' && success(res);
    typeof complete === 'function' && complete(res);
    return Promise.resolve(res);
}
function getStorageSync(key) {
    if (typeof key !== 'string') {
        console.error(getParameterError({
            name: 'getStorage',
            correct: 'String',
            wrong: key
        }));
        return;
    }
    var res = getItem(key);
    if (res.result)
        return res.data;
    return '';
}
function getItem(key) {
    var item;
    try {
        item = JSON.parse(localStorage.getItem(key));
    }
    catch (e) { }
    // 只返回使用 Taro.setStorage API 存储的数据
    if (item && typeof item === 'object' && item.hasOwnProperty('data')) {
        return { result: true, data: item.data };
    }
    else {
        return { result: false };
    }
}
function getStorageInfo(options) {
    if (options === void 0) { options = {}; }
    var success = options.success, complete = options.complete;
    var res = { errMsg: 'getStorageInfo:ok' };
    var info = getStorageInfoSync();
    Object.assign(res, info);
    success && success(res);
    complete && complete(res);
    return Promise.resolve(res);
}
function getStorageInfoSync() {
    var res = {};
    res.keys = Object.keys(localStorage);
    res.limitSize = null;
    res.currentSize = null;
    return res;
}
function removeStorage(options) {
    // options must be an Object
    var isObject = shouleBeObject(options);
    if (!isObject.res) {
        var res_3 = { errMsg: "removeStorage" + isObject.msg };
        console.error(res_3.errMsg);
        return Promise.reject(res_3);
    }
    var key = options.key, success = options.success, fail = options.fail, complete = options.complete;
    var res = { errMsg: 'removeStorage:ok' };
    if (typeof key !== 'string') {
        res.errMsg = getParameterError({
            name: 'removeStorage',
            para: 'key',
            correct: 'String',
            wrong: key
        });
        console.error(res.errMsg);
        typeof fail === 'function' && fail(res);
        typeof complete === 'function' && complete(res);
        return Promise.reject(res);
    }
    removeStorageSync(key);
    success && success(res);
    complete && complete(res);
    return Promise.resolve(res);
}
function removeStorageSync(key) {
    if (typeof key !== 'string') {
        console.error(getParameterError({
            name: 'removeStorage',
            correct: 'String',
            wrong: key
        }));
        return;
    }
    localStorage.removeItem(key);
}
function clearStorage() {
    localStorage.clear();
}
export { setStorage, setStorageSync, getStorage, getStorageSync, getStorageInfo, getStorageInfoSync, removeStorage, removeStorageSync, clearStorage, clearStorage as clearStorageSync };
